<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="ICARDA" />


<title>FAGE Analysis Pipeline (ver 0.6.1)</title>

<script src="FAGE_Analysis_Pipeline_files/header-attrs-2.21/header-attrs.js"></script>
<script src="FAGE_Analysis_Pipeline_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="FAGE_Analysis_Pipeline_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="FAGE_Analysis_Pipeline_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="FAGE_Analysis_Pipeline_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="FAGE_Analysis_Pipeline_files/navigation-1.1/tabsets.js"></script>
<script src="FAGE_Analysis_Pipeline_files/navigation-1.1/codefolding.js"></script>
<link href="FAGE_Analysis_Pipeline_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/highlightjs-9.12.0/highlight.js"></script>
<script src="FAGE_Analysis_Pipeline_files/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="FAGE_Analysis_Pipeline_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/datatables-binding-0.27/datatables.js"></script>
<link href="FAGE_Analysis_Pipeline_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="FAGE_Analysis_Pipeline_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/jszip-1.12.1/jszip.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/pdfmake-1.12.1/pdfmake.js"></script>
<script src="FAGE_Analysis_Pipeline_files/pdfmake-1.12.1/vfs_fonts.js"></script>
<link href="FAGE_Analysis_Pipeline_files/dt-ext-buttons-1.12.1/css/buttons.dataTables.min.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/dt-ext-buttons-1.12.1/js/dataTables.buttons.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/dt-ext-buttons-1.12.1/js/buttons.html5.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/dt-ext-buttons-1.12.1/js/buttons.colVis.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/dt-ext-buttons-1.12.1/js/buttons.print.min.js"></script>
<link href="FAGE_Analysis_Pipeline_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<script src="FAGE_Analysis_Pipeline_files/plotly-binding-4.10.1/plotly.js"></script>
<script src="FAGE_Analysis_Pipeline_files/typedarray-0.1/typedarray.min.js"></script>
<link href="FAGE_Analysis_Pipeline_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="FAGE_Analysis_Pipeline_files/plotly-main-2.11.1/plotly-latest.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">FAGE Analysis Pipeline (ver 0.6.1)</h1>
<h4 class="author">ICARDA</h4>
<h4 class="date">At 9:57 AM on Wednesday, 6 September, 2023</h4>

</div>


<a href="https://github.com/icarda-git/FAGE/" target="_blank" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a>
<style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
<style> body {text-align: justify} </style>
<div id="analytical-flow-of-pipeline" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Analytical Flow of
Pipeline</h1>
<p>Multi-environmental trial (MET) analyses are critical for plant
breeding programs. In MET group of genotypes are evaluated in several
environments (sites, years, or treatments) in order to assess their
levels of consistency (or stability) of the performance across these
different environments. This lack of stability is known as
genotype-by-environment (GxE) interaction and is the result of specific
gene responses to the particular conditions of a given environment.</p>
<p><strong>Single- versus Two-stage Analysis</strong></p>
<p>For the analyses of MET, ideally all trials should be evaluated by
fitting a unique complete model. This is known as <strong>single-stage
analysis</strong>. In here, the original phenotypic measurements from
all trials are modeled together and genetic, non-genetic and residuals
variances for each trial are estimated. This process can be challenging
and often there are computational complications, mostly originating by
the large number of variance components to simultaneously estimate.</p>
<p>An alternative to the above approach is to perform a
<strong>two-stages analysis</strong>. This is divided into two steps. In
the step 1 each trial is first analyzed individually. Here, all relevant
factors, covariates and spatial components are considered, and once a
final model is selected, adjusted means (BLUEs) are obtained for each of
the genotypes in the trial together with its weights (further details
later). Then, <strong>in the step 2</strong>, all individual trial
analyses are combined and used to fit a single linear mixed-model (LMM)
that considers weights.</p>
<p>Both the single-stage and two-stages analyses have advantages and
disadvantages. The most important advantage of the two-stages analysis
is that it allows considering any relevant design or statistical
elements for each trial. This includes different aspects such as spatial
analyses, replication, etc. It is common for a single-stage analysis to
have too much complexity making it difficult or near impossible to fit,
and often some model simplifications are required (e.g., dropping some
model terms). These simplifications can translate into important losses
of accuracy.</p>
<p>Similarly, in the two-stages analyses, there is some loss of
information, as the parameter estimation is done in two steps. However,
this separation, allows for the first step to have as much complexity as
desired, and often better information is obtained from this stage.</p>
<p>There is another important advantage of the two-stages analysis, and
this is related to developing operational analytical pipelines. A
single-stage analysis requires all the raw data available for the
analysis at the moment of fitting the model. In contrast, the two-stages
analyses the first step can be executed in advance and this information
can be then retrieved to fit a second step with the desired trials and
genotypes. This has also the advantage that the adjusted means (and
weights) from the first step can be used for other studies such as
genomic prediction (GP) and/or genome-wide association studies
(GWAS).</p>
<p><strong>Steps of the Two-stage Analysis</strong></p>
<p>The focus on this pipeline will be on the use of two-stage analyses.
As indicated earlier, in the first step, the raw phenotypic data of each
trial is analyzed. The idea is to use all available statistical tools to
fit each of these trials providing the largest possible accuracy on the
genotype’s predictions. In the current pipeline, each site is been
evaluated considering genotype as a random effect (allowing for the
estimation of heritability), and several model terms are considered to
control for potential spatial heterogeneity. These include elements such
as autoregressive errors or order 1 (AR1) for the rows and columns,
polynomials or spline to model trends, random row or column effects
nested within replicate also to model trends.</p>
<p>The procedure evaluates several of these models and according to a
given criteria (AIC, BIC, heritability) where the best fitted model is
selected. If a low heritability is found (threshold specified by the
user) then the trials are not considered in the second step of this
analysis. Otherwise, the fitting proceeds. Here, all variance component
estimates are fixed to their estimates, and the model is refitted with
the genotype effect considered as a fixed effect. From here, adjusted
means (or BLUE means, or predictions) are obtained for each of the
genotypes together with their weights. Weights are obtained as the
diagonal of the inverse of the variance-covariance matrix of adjusted
mean estimates (or predictions).</p>
<p>It is important to note, that the current pipeline (workflow) also
considers a careful exploration of each individual trial fit, where
residual plots are checked, outliers are detected (and eliminated) in a
semi-automatic way, and several other statistics are available (such as
coefficient of variation, CV%) in order to assess the quality of each of
these model fits. These fits should be checked carefully, as the
predictions obtained will be stored and possibly used for many years and
for different statistical evaluations.</p>
<p>For the second step, information from each of the individual fits is
collected. This is an important aspect of the workflow, as different
sets of trials can be selected for different objectives. For example,
only trials from the current year might be evaluated, but also trials
from, for example drought years, might be combined into a single
two-stage MET analysis.</p>
<p>One important aspect to consider in this stage before proceeding is
to check connectivity between trials by auditing the collected data. The
MET analysis will not converge if no genotypes are common between some
pairs of trials, or if only very few genotypes are common between
trials. As a rule of thumb, we recommend having at least 5 genotypes in
common. However, note that important levels of unbalanced are often
tolerated by LMMs (particularly in factor analytic structures), but
still some connectivity is still required.</p>
<p>After this audit, it is possible to fit the LMM with complex
structures. Here, the current workflow considers the structures of
single correlation and common variance (CORV), single correlation and
heterogeneous variances (CORH), and several factor analytic (FA)
structures.</p>
<p>The focus of the current pipeline is mainly on the FA structure as
this has several advantages such as: 1) greatest model flexibility with
approximations to different correlations and heterogeneous variances, 2)
straightforward interpretation of some of the components (loadings,
specific, etc.), 3) allows for estimation of some correlations even
under poor connectivity, 4) less convergency issues than other
structures, such as CORGH and US. In addition, this structure allows for
a better understanding of the patterns of GxE and some selection of
individuals for underlying factors that affect the response, and
therefore that induces the observed pattern of GxE effects found in the
data.</p>
<p>Further details and illustrations of these steps are presented in the
example presented here.</p>
</div>
<div id="getting-started" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Getting Started</h1>
<p>There are several R libraries that are used for statistical analyses
in the present workflow. The following code specifies which libraries
are required, and these will be installed if not available.</p>
<ul>
<li><p>ASReml-R: requires a commercial license to be installed (for
further information please check: <a
href="https://www.vsni.co.uk/software/asreml-r">this</a>)</p></li>
<li><p>QBMS: requires access (login) to the working databases from the
BMS system.</p></li>
</ul>
<p>The specifications of the analyses are presented in the file
config.in that assists on the data manipulations and model fit.</p>
<pre class="r"><code>required_packages &lt;- c(&quot;QBMS&quot;, &quot;asreml&quot;, &quot;ggplot2&quot;, &quot;lattice&quot;, &quot;gridExtra&quot;, &quot;dplyr&quot;, &quot;plyr&quot;, &quot;ggrepel&quot;, &quot;tidyr&quot;,&quot;ggfortify&quot;, &quot;missMDA&quot;, &quot;ggpubr&quot;, &quot;pheatmap&quot;, &quot;kableExtra&quot;, &quot;DT&quot;, &quot;configr&quot;, &quot;remotes&quot;, &quot;plotly&quot;)

not_installed &lt;- required_packages[!(required_packages %in% installed.packages()[ , &quot;Package&quot;])]
if(length(not_installed)) install.packages(not_installed, repos = &quot;http://cran.us.r-project.org&quot;)

#if (!require(QBMS)) remotes::install_github(&quot;icarda-git/QBMS&quot;)
if (!require(bsselectR)) remotes::install_github(&quot;walkerke/bsselectR&quot;)

# stage-1 required libraries
library(asreml)</code></pre>
<pre><code>## Online License checked out Wed Sep  6 10:57:46 2023</code></pre>
<pre class="r"><code>library(ggplot2)
library(lattice)
library(gridExtra)
library(dplyr)
library(plyr)
library(QBMS)
library(plotly)

# stage-2 extra required libraries
library(ggrepel)
library(tidyr)     # required for reshaping long to wide table of data
library(ggfortify) # required by the ./lib/GBiplot.R
library(missMDA)   # required by the ./lib/GBiplot.R
library(ggpubr)    # get_legend
library(pheatmap)  # pretty heatmap!

# r markdown required libraries
library(kableExtra)
library(DT)
library(configr)
library(bsselectR)

# stage-1 required includes
source(&quot;./lib/bms_import.R&quot;)
source(&quot;./lib/detect_outliers.R&quot;)
source(&quot;./lib/mcomp_asreml.R&quot;)
source(&quot;./lib/gral_stats.R&quot;)
source(&quot;./lib/gral_single_wgt.R&quot;)
source(&quot;./lib/spatial_selector.R&quot;)
source(&quot;./lib/fieldmap.R&quot;)

# stage-2 required includes
source(&quot;./lib/auditMET.R&quot;)
source(&quot;./lib/stageMET.R&quot;)
source(&quot;./lib/GBiplot.R&quot;)
source(&quot;./lib/fa_plots.R&quot;)

# get model definitions
load(file = &quot;./lib/MODLIST.Rda&quot;)

# remove previouse analysis outputs/plots
unlink(&quot;outputs/*&quot;, force = TRUE)
unlink(&quot;plots/*&quot;, force = TRUE)

config_file &lt;- &quot;config.ini&quot;

if (!file.exists(config_file)) {
  warning(paste(config_file, &quot;not exists!&quot;))
  knitr::knit_exit()
}

config &lt;- read.config(file = config_file)

cat(readLines(config_file), sep = &#39;\n&#39;)</code></pre>
<pre><code>## [default]
## study.name  = Study
## trait.name  = GY_Calc_kgha
## source.bms  = FALSE
## both.stages = TRUE
## 
## [bms]
## ; define extra params required by BMS API to retrieve the data
## url     = https://bms.icarda.org/ibpworkbench/
## crop    = wheat
## program = Wheat International Nurseries
## 
## [outliers]
## threshold.start = 4
## threshold.stop  = 3
## threshold.step  = 0.25
## 
## [met]
## heritVC.limit = 0.10
## Pvalue.limit  = 0.10
## excluded.loc  = loc_17, loc_09
## excluded.geno = geno_02, geno_11, geno_16, geno_20
## 
## best.fa =</code></pre>
</div>
<div id="import-and-loading-data" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Import and Loading
Data</h1>
<p>In this section we will be importing raw data directly from BMS.
However, a standalone file with raw data can also be read and
accessed.</p>
<pre class="r"><code># the trial and trait names (e.g., IDYT39 and GY_Calc_kgha respectively)
study.name &lt;- config$default$study.name
trait.name &lt;- config$default$trait.name</code></pre>
<div id="import-from-bms" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Import from BMS</h2>
<p>If you are importing data from BMS (i.e., <code>source.bms</code> is
TRUE), then <code>study.name</code> (e.g., “Study”) should exist in the
following <code>bms.program</code> in your <code>bms.crop</code>. Also,
the <code>trait.name</code> (e.g., “GY_Calc_kgha”) should match your BMS
ontology and has observed data. Please check these settings in the
associated config file <a href="./config.ini">here</a>.</p>
<p>This analysis pipeline is using the <a
href="https://icarda-git.github.io/QBMS/">QBMS</a> R package to query
the <a href="https://bmspro.io/">Breeding Management System</a> database
(using <a href="https://brapi.org/">BrAPI</a> calls) and retrieves
experiments data directly into R statistical analyzing environment.</p>
<blockquote>
<p>You will get a login pop-up window, use the your BMS credentials</p>
</blockquote>
<pre class="r"><code># BMS is your data source?
source.bms &lt;- (config$default$source.bms == &quot;TRUE&quot;)

# define extra params required by BMS API to retrieve the data
bms.url     &lt;- config$bms$url
bms.crop    &lt;- config$bms$crop
bms.program &lt;- config$bms$program</code></pre>
</div>
<div id="import-from-other-sources" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Import from Other
Sources</h2>
<p>If you are reading data directly, then two CSV files are required
here in the same directory of this script named as the following:</p>
<ul>
<li><p>Study_germplasm.csv for your list of germplasms, it should
contain two columns <code>germplasmName</code> and <code>check</code> (0
for test entries and 1 for check/control entries).</p></li>
<li><p>Study_GY_Calc_kgha_data.csv for your trials data. Mandatory
columns are <code>trial</code>, <code>gen</code> (genotypes, they should
match with the ones in the previous germplasm list), <code>block</code>
(replication), <code>ibk</code> (incomplete blocks), <code>col</code>
and <code>row</code> (field layout), and <code>resp</code> (trait
response).</p></li>
</ul>
<blockquote>
<p>If no block (replication) is considered in the designs, this should
be filled with a 1 in the dataset. Also, if no ibk (incomplete blocks),
column or row coordinates are available, these should be provided as NA
in the dataset.</p>
</blockquote>
<p>Get the name of required input files and check if they exist:</p>
<pre class="r"><code>if(source.bms){
  # data will extracted directly from BMS server and save in intermediate csv files
  files &lt;- bms_import(study.name, trait.name, bms.crop, bms.program, bms.url)
  geno_file &lt;- files$germplasm
  data_file &lt;- files$data
}else{
  # setup the expected names 
  geno_file &lt;- paste0(study.name, &quot;_germplasm.csv&quot;)
  data_file &lt;- paste0(study.name, &quot;_&quot;, trait.name, &quot;_data.csv&quot;)
}

if (!file.exists(data_file)) {
  warning(paste(data_file, &quot;not exists!&quot;))
  knitr::knit_exit()
}
if (!file.exists(geno_file)) {
  warning(paste(geno_file, &quot;not exists!&quot;))
  knitr::knit_exit()
}</code></pre>
</div>
<div id="check-for-mandatory-columns" class="section level2"
number="3.3">
<h2><span class="header-section-number">3.3</span> Check for Mandatory
Columns</h2>
<p>The following code is an internal check that the required columns are
provided in their respective data frames.</p>
<pre class="r"><code># load data 
data &lt;- read.csv(data_file)
germplasm.data &lt;- read.csv(geno_file)

if(&quot;studyId&quot; %in% colnames(data)) {
  bms.studyId  &lt;- data$studyId[1]
  data$studyId &lt;- NULL
}

# QC: check for mandatory columns
oops &lt;- setdiff(c(&quot;trial&quot;, &quot;gen&quot;, &quot;block&quot;, &quot;ibk&quot;, &quot;col&quot;, &quot;row&quot;, &quot;resp&quot;), colnames(data))
if(length(oops) &gt; 0) {
  warning(paste(&quot;Can&#39;t find the following mandatory columns in your data file&quot;, 
                data_file, &quot;\n&quot;, paste(oops,  collapse = &quot;, &quot;)))
  knitr::knit_exit()
}

oops &lt;- setdiff(c(&quot;germplasmName&quot;, &quot;check&quot;), colnames(germplasm.data))
if(length(oops) &gt; 0) {
  warning(paste(&quot;Can&#39;t find the following mandatory columns in your germplasm file&quot;, 
                geno_file, &quot;\n&quot;, paste(oops, collapse = &quot;, &quot;)))
  knitr::knit_exit()
}</code></pre>
</div>
<div id="data-pre-processing" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Data
Pre-processing</h2>
<p>In this section, some conversions of columns within a data frame are
performed to be ready for the downstream analyses (e.g., convert to
factor, sorting, and unique identifier).</p>
<pre class="r"><code># make sure that design and experimental treatments are factors, and response is numeric
data$gen   &lt;- factor(data$gen)
data$trial &lt;- factor(data$trial)
data$block &lt;- factor(data$block)
data$ibk   &lt;- factor(data$ibk)
data$col   &lt;- factor(data$col)
data$row   &lt;- factor(data$row)
data$resp  &lt;- as.numeric(as.character(data$resp))

# you MUST sort your data into the order you specify in the rcov/residuals term
data &lt;- data[with(data, order(trial, row, col)),]

# create a unique record identifier
data$id &lt;- 1:dim(data)[1]</code></pre>
</div>
<div id="extract-check-entries" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Extract Check
Entries</h2>
<p>Extract the names of check entries and tag them in the data file:</p>
<blockquote>
<p>It is important to verify that these entries are the ones defined as
checks (or controls).</p>
</blockquote>
<pre class="r"><code># germplasm list has a column refers if any entry is test(0) or check(1)
check.name &lt;- germplasm.data$germplasmName[germplasm.data$check == 1]

# create check column (handle weird cases like extra spaces and/or uppercase/lowercase letters!)
data$check &lt;- ifelse(gsub(&quot; &quot;,&quot;&quot;,tolower(data$gen)) %in% gsub(&quot; &quot;,&quot;&quot;,tolower(check.name)), 1, 0)</code></pre>
<blockquote>
<p>Check entries are: geno_02, geno_11, geno_16, geno_20</p>
</blockquote>
</div>
<div id="define-experimental-design" class="section level2"
number="3.6">
<h2><span class="header-section-number">3.6</span> Define Experimental
Design</h2>
<p>Define the type of experiment design for each trial and assess the
availability of spatial information. The available designs in this
pipeline are:</p>
<ul>
<li>ALPHA: alpha-lattice design.</li>
<li>RCBD: randomized complete block design.</li>
<li>AUGMENTED: RCBD augmented design in a RCBD.</li>
<li>PREP: partially-replicated design.</li>
</ul>
<pre class="r"><code># create a new data.frame to report all issues related to the row/col data if there is any
spatial.issues &lt;- data.frame(&quot;trial&quot; = character(), &quot;issue&quot; = character(), 
                             &quot;row&quot; = character(), &quot;col&quot; = character())

# create a new data.frame list experiment design of each trial (ALPHA, RCBD, AUGMENTED, PREP)
trial.type &lt;- data.frame(&quot;trial&quot; = character(), stringsAsFactors = FALSE)

#***************************************
#* Design       has.block     has.ibk
#* 
#* CRD          FALSE         FALSE
#* AUGMENTED    FALSE         TRUE
#* RCBD         TRUE          FALSE
#* ALPHA        TRUE          TRUE
#***************************************

# can guess experiment design depends on presence/absence of block and ibk (i.e., rep and block) design factors
for(i in 1:nlevels(data$trial)){
  site.data &lt;- subset(data, trial == levels(data$trial)[i])
  trial.type[i, &quot;trial&quot;]  &lt;- levels(data$trial)[i]

  has.block &lt;- !anyNA(site.data$block) &amp;&amp; length(unique(site.data$block)) &gt; 1
  has.ibk   &lt;- !anyNA(site.data$ibk)   &amp;&amp; length(unique(site.data$ibk))   &gt; 1

  trial.type[i, &quot;design&quot;] &lt;- ifelse(has.block, 
                                    ifelse(has.ibk, &quot;ALPHA&quot;, &quot;RCBD&quot;), 
                                    ifelse(has.ibk, &quot;AUGMENTED&quot;, &quot;CRD&quot;))

  has.spatial &lt;- !(anyNA(site.data$row) | anyNA(site.data$col))
  
  if (has.spatial) {
    # assure the row/column data contiguity in the full field dimensions
    contiguous.data &lt;- expand.grid(unique(site.data$row), unique(site.data$col))
    colnames(contiguous.data) &lt;- c(&quot;row&quot;, &quot;col&quot;)
    missing.coords &lt;- setdiff(contiguous.data, site.data[,c(&quot;row&quot;, &quot;col&quot;)])
    
    if (!empty(missing.coords)) {
      missing.coords$issue &lt;- &quot;missing row x col&quot;
      missing.coords$trial &lt;- trial.type[i, &quot;trial&quot;]
      spatial.issues &lt;- rbind(spatial.issues, missing.coords)
    }
    
    # check if there is any duplicate row/col coordinates
    duplicated.coords &lt;- site.data[duplicated(site.data[,c(&quot;row&quot;, &quot;col&quot;)]),]
    
    if (!empty(duplicated.coords)) {
      duplicated.coords &lt;- duplicated.coords[,c(&quot;row&quot;, &quot;col&quot;)]
      duplicated.coords$issue &lt;- &quot;duplicated row x col&quot;
      duplicated.coords$trial &lt;- trial.type[i, &quot;trial&quot;]
      
      spatial.issues &lt;- rbind(spatial.issues, duplicated.coords)
    }
    
    trial.type[i, &quot;has.spatial&quot;] &lt;- paste0(&quot;Yes (&quot;, length(unique(site.data$row)), &quot;x&quot;, 
                                          length(unique(site.data$col)), &quot; row/col)&quot;)
    
  } else {
    trial.type[i, &quot;has.spatial&quot;] &lt;- &quot;NO&quot;
  }
}

create_dt(trial.type)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-9fffdaaa1d332772ef73" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-9fffdaaa1d332772ef73">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"],["loc_01","loc_02","loc_03","loc_04","loc_05","loc_06","loc_07","loc_08","loc_09","loc_10","loc_11","loc_12","loc_13","loc_14","loc_15","loc_16","loc_17","loc_18","loc_19","loc_20","loc_21"],["ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA"],["Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)","Yes (4x12 row/col)"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>trial<\/th>\n      <th>design<\/th>\n      <th>has.spatial<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>if (!empty(spatial.issues)) {
  warning(&quot;has some issues in spatial data integrity!&quot;)
  create_dt(spatial.issues)
}</code></pre>
<div class="alert alert-warning">
<p><strong>Warning</strong> has some issues in spatial data
integrity!</p>
</div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-28b1ac4d7028b16bbc2e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-28b1ac4d7028b16bbc2e">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42"],["4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2","4","2"],["11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12","11","12"],["missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col","missing row x col"],["loc_01","loc_01","loc_02","loc_02","loc_03","loc_03","loc_04","loc_04","loc_05","loc_05","loc_06","loc_06","loc_07","loc_07","loc_08","loc_08","loc_09","loc_09","loc_10","loc_10","loc_11","loc_11","loc_12","loc_12","loc_13","loc_13","loc_14","loc_14","loc_15","loc_15","loc_16","loc_16","loc_17","loc_17","loc_18","loc_18","loc_19","loc_19","loc_20","loc_20","loc_21","loc_21"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>row<\/th>\n      <th>col<\/th>\n      <th>issue<\/th>\n      <th>trial<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="field-maps" class="section level2" number="3.7">
<h2><span class="header-section-number">3.7</span> Field Maps</h2>
<p>The following code provides a field map to assess the spatial
distribution of the response variable for each trial with available
spatial information. This is recommended in order to identify potential
outliers, trait patterns in the field, potential post-blocking, etc.</p>
<pre class="r"><code># draw fieldmaps!
field.plots &lt;- list()
plot.names  &lt;- list()

for(i in 1:nlevels(data$trial)){
  site.data &lt;- subset(data, trial == levels(data$trial)[i])
  if (anyNA(site.data$row) | anyNA(site.data$col)) next
  field.plots[[i]] &lt;- fieldmap(x=site.data, column=&quot;resp&quot;, textColumn=&quot;gen&quot;, asFactor=F, 
                               rows=&quot;row&quot;, ranges=&quot;col&quot;, borderColumns=&quot;ibk&quot;, 
                               main=paste(levels(data$trial)[i], trait.name))
  
  plot.names[[i]] &lt;- levels(data$trial)[i]
  png(paste0(&quot;plots/fieldmap_&quot;, plot.names[[i]], &quot;_&quot;, trait.name, &quot;.png&quot;))
  print(field.plots[[i]])
  dev.off()
}

if (length(plot.names) != 0) {
  field.plots[sapply(field.plots, is.null)] &lt;- NULL
  plot.names[sapply(plot.names, is.null)] &lt;- NULL
  
  pages &lt;- marrangeGrob(grobs = field.plots, nrow=1, ncol=1)
  
  page.width  &lt;- 2 * max(as.numeric(data$col))
  page.height &lt;- 5 * max(as.numeric(data$row))
  
  ggsave(paste0(&quot;outputs/&quot;, study.name, &quot;_&quot;, trait.name, &quot;_Stage-1_Fieldmap.pdf&quot;), 
         plot = pages, width = page.width, height = page.height, dpi = 600, units = &quot;cm&quot;, limitsize = FALSE)
  
  select.plot &lt;- paste0(&quot;plots/fieldmap_&quot;, plot.names, &quot;_&quot;, trait.name, &quot;.png&quot;)
  names(select.plot) &lt;- plot.names
  bsselect(select.plot, type = &quot;iframe&quot;, live_search = TRUE, show_tick = TRUE, elementId = &quot;fieldmaps&quot;)$prepend[[1]]$html
} else {
  message(&quot;There is no spatial data!&quot;)
}</code></pre>
<html>
<select id="IPCTUbNIrP" class="selectpicker" data-dropdown-align-right="false" data-dropup-auto="true" data-header="false" data-live-search="true" data-live-search-style="contains" data-show-tick="true" data-width="false" data-size="auto"><option value="plots/fieldmap_loc_01_GY_Calc_kgha.png">loc_01</option>
<option value="plots/fieldmap_loc_02_GY_Calc_kgha.png">loc_02</option>
<option value="plots/fieldmap_loc_03_GY_Calc_kgha.png">loc_03</option>
<option value="plots/fieldmap_loc_04_GY_Calc_kgha.png">loc_04</option>
<option value="plots/fieldmap_loc_05_GY_Calc_kgha.png">loc_05</option>
<option value="plots/fieldmap_loc_06_GY_Calc_kgha.png">loc_06</option>
<option value="plots/fieldmap_loc_07_GY_Calc_kgha.png">loc_07</option>
<option value="plots/fieldmap_loc_08_GY_Calc_kgha.png">loc_08</option>
<option value="plots/fieldmap_loc_09_GY_Calc_kgha.png">loc_09</option>
<option value="plots/fieldmap_loc_10_GY_Calc_kgha.png">loc_10</option>
<option value="plots/fieldmap_loc_11_GY_Calc_kgha.png">loc_11</option>
<option value="plots/fieldmap_loc_12_GY_Calc_kgha.png">loc_12</option>
<option value="plots/fieldmap_loc_13_GY_Calc_kgha.png">loc_13</option>
<option value="plots/fieldmap_loc_14_GY_Calc_kgha.png">loc_14</option>
<option value="plots/fieldmap_loc_15_GY_Calc_kgha.png">loc_15</option>
<option value="plots/fieldmap_loc_16_GY_Calc_kgha.png">loc_16</option>
<option value="plots/fieldmap_loc_17_GY_Calc_kgha.png">loc_17</option>
<option value="plots/fieldmap_loc_18_GY_Calc_kgha.png">loc_18</option>
<option value="plots/fieldmap_loc_19_GY_Calc_kgha.png">loc_19</option>
<option value="plots/fieldmap_loc_20_GY_Calc_kgha.png">loc_20</option>
<option value="plots/fieldmap_loc_21_GY_Calc_kgha.png">loc_21</option></select>
<iframe src="plots/fieldmap_loc_01_GY_Calc_kgha.png" frameborder="0" height="500" width="100%" id="x8aTmVIXtp"></iframe>
<script>$(document).ready(function(){
                   $("#IPCTUbNIrP").change(function(){
                   $("#x8aTmVIXtp").attr("src",$(this).val());

                   });

    });</script>
</html>
<blockquote>
<p><em>Check your field maps <a
href="./outputs/Study_GY_Calc_kgha_Stage-1_Fieldmap.pdf">here</a></em></p>
</blockquote>
</div>
</div>
<div id="single-environment-analysis" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Single-Environment
Analysis</h1>
<p>This is the first step of the two-stages analyses where each trial
(or environment) is evaluated individually to obtain the ‘best’ possible
fit for the trial under study.</p>
<div id="basic-distribution-plots" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Basic Distribution
Plots</h2>
<p>These plots highly potential differences in the data that might
affect the analyses and should alert the users. These plots assist
identifying trials/individuals with large variations or large
differences between replications/blocks.</p>
<pre class="r"><code>par(mar = c(10, 4, 2, 2) + 0.1)
xyplot(resp ~ gen | trial, groups = block, data = data, 
       scales = list(draw = FALSE), strip = strip.custom(par.strip.text = list(cex = 0.8)))</code></pre>
<p><img src="FAGE_Analysis_Pipeline_files/figure-html/distribution_plots-1.png" width="1800" /></p>
<pre class="r"><code>plot_ly(data, y = ~resp, x = ~trial, type = &quot;box&quot;, width = 800, height = 480)</code></pre>
<div id="htmlwidget-49ab13c6f45b3b24b011" style="width:800px;height:480px;" class="plotly html-widget "></div>
<script type="application/json" data-for="htmlwidget-49ab13c6f45b3b24b011">{"x":{"visdat":{"36b43166e0":["function () ","plotlyVisDat"]},"cur_data":"36b43166e0","attrs":{"36b43166e0":{"y":{},"x":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"box"}},"layout":{"width":800,"height":480,"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"trial","type":"category","categoryorder":"array","categoryarray":["loc_01","loc_02","loc_03","loc_04","loc_05","loc_06","loc_07","loc_08","loc_09","loc_10","loc_11","loc_12","loc_13","loc_14","loc_15","loc_16","loc_17","loc_18","loc_19","loc_20","loc_21"]},"yaxis":{"domain":[0,1],"automargin":true,"title":"resp"},"hovermode":"closest","showlegend":false},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"fillcolor":"rgba(31,119,180,0.5)","y":[1275,1295,1620,1663,1015,1450,1465,1943,1285,950,1283,1168,1380,1413,1273,1273,1525,1078,1575,1553,1643,1225,910,1648,1740,1330,1800,1515,1250,1205,1218,1310,1165,1193,1498,1208,1360,1300,1438,1253,1573,1795,1260,1195,1505,1018,3600,2200,3500,3000,4000,3100,5200,6400,5000,6200,4200,6200,4000,3600,4600,3400,4000,4300,4000,4600,4000,5500,2800,5000,5000,2500,2800,3200,2000,2200,2800,3200,2200,2000,2400,1200,1800,2200,2000,2050,1600,2000,1000,900,1000,4200,2091,812,2254,1000,2045,1600,1094,1716,1670,1666,834,2355,1940,1973,1630,969,1094,1184,1853,1557,1831,1591,1125,825,2011,1650,1894,1570,1013,1296,1285,1660,1263,799,1541,1708,1577,1928,1908,1714,1410,1698,939,1913,1850,2128,6860,8693,8863,8620,8340,8617,9217,6973,8417,8070,8517,6363,6270,9410,8053,8043,8220,6360,6793,6637,7003,6513,6140,9370,9423,10003,9503,6013,6600,8983,10097,8417,6490,9530,5883,8130,10217,6943,7200,6193,9647,8803,9893,10440,5687,9747,2890,2372,2503,2162,2997,2395,2188,2712,1688,1752,1752,1437,1775,2743,2455,2373,2793,2548,2082,1967,2052,968,1840,2602,1485,1085,2342,2448,2453,1960,1857,2967,2252,2323,3157,2502,1900,1957,2280,1530,1337,2220,2372,3258,2595,2025,3167,4500,4333,4667,5333,4083,4167,5333,4833,4667,4000,4833,5667,5333,4000,3833,4167,4500,4000,5833,5800,4667,4333,3333,4167,4167,4833,5500,4333,4333,5333,5167,4833,3833,5000,5167,5000,3833,4133,4500,4333,3667,5500,5667,4667,4167,9560,5208,7176,6134,8681,7292,6319,7523,5833,7917,5903,7639,7940,7870,8218,7593,8912,6296,7407,5856,6296,5347,7616,8611,8171,7176,9144,7222,7940,8403,6713,8056,6134,6296,8264,7917,6759,6759,7685,6759,7361,8403,5139,6366,7292,7986,7158,7172,6806,6658,8072,7853,7744,8106,7011,8439,8403,8708,6956,8878,9242,7128,6550,8261,7903,7306,6164,8172,7592,7231,7792,8150,8675,9717,9236,8828,7522,6886,6089,8283,8556,7175,6764,7553,7761,7761,7789,7439,7694,9389,9219,8611,980,1020,840,910,990,720,820,870,1120,1090,1090,900,910,900,1020,1100,1080,1110,1080,1110,1140,990,920,1110,960,1000,1040,1100,1040,770,850,1120,1110,1100,1010,990,990,920,850,1210,960,910,1040,900,950,940,1542,1525,1692,1283,1217,2025,1875,1575,1725,3283,3025,2517,2817,3158,3150,2200,2267,2142,1683,1992,1642,2075,3025,3200,1950,2083,2025,1917,1850,1325,1292,1700,1450,2767,3117,3317,2158,2067,2475,1292,1417,1558,1858,1608,1417,1317,1008,1324,1488,1273,1086,1831,1416,1629,1116,1537,1884,898,1733,1429,1708,1429,1363,1473,1351,1618,1543,2096,1329,1025,1335,1492,1273,1635,1237,1680,1659,910,1069,1878,910,1512,1543,1702,904,1657,1653,1688,1600,1520,1514,1329,1316,1224,2026,2053,2132,2158,2211,3342,1737,1737,1553,1982,2368,1955,1729,1316,2024,1853,1645,1758,2092,2139,1626,2737,2368,1921,2474,2500,2184,1724,2539,2263,2026,1868,2684,1579,1579,1789,2974,2158,2053,2579,1842,1868,1618,1474,5798,3887,5899,4869,6846,6126,7769,6299,6019,6351,5357,6280,5661,5637,6369,5417,5893,6315,6315,5667,5720,4815,6280,5637,7101,5292,6839,5821,5952,6202,5804,6738,5512,4762,5964,6089,5893,5911,7714,5869,5530,5625,4429,5071,5780,5185,3992,3950,4208,3217,2500,2600,3100,3850,2675,3167,3817,2958,3750,3238,5033,3792,3683,3425,3092,3333,2250,4117,2083,3642,3350,2250,2825,3183,2033,2625,2908,3042,2850,5475,5583,4483,4783,4167,3558,2317,3505,3525,4133,3550,2575,4158,2706,1740,3060,1540,1980,2300,3120,2840,2840,3180,2520,2460,2400,3660,4340,4580,3220,2800,2860,1420,1900,2660,2120,2737,2420,3080,1840,2420,2420,2580,4400,3780,3560,3040,1860,2220,3020,2920,2560,3100,3140,3080,1400,1840,2560,2980,10167,8595,10690,7833,10500,8500,10762,9000,11190,8976,9500,11190,9905,11048,10833,8952,10214,10167,9976,10786,8738,9571,9262,11381,7333,10262,10524,9905,8714,9571,6524,9405,9048,9310,11167,10286,10167,10214,10595,10262,9810,9762,8857,8143,9976,10071,3390,3133,3508,3757,3621,2623,3624,3380,2627,3643,2751,3320,2942,3306,3807,3885,3138,3457,3355,4176,3767,3458,2552,3081,3919,3005,3762,3188,3661,2867,3780,3444,3636,3146,3409,2571,3229,3393,4310,4336,3244,4163,3097,3743,3617,2588,3900,2008,2375,3258,2208,2025,2500,2950,2242,2325,2667,2958,2675,2500,1833,2883,3617,3608,2733,2167,3508,3003,2742,2892,2958,3025,2500,3100,3408,1667,3692,3383,3600,2667,1750,2567,2317,2495,2792,2408,2767,2908,2008,3900,3392,2758,9778,8629,9185,7814,9203,10295,8710,9047,9973,10175,9985,9923,8438,9124,9560,9860,9195,9161,9073,8998,8580,8285,9396,8795,9023,7960,10083,9625,8995,9763,7214,8961,8378,8960,9503,9196,9473,7893,8300,8055,8924,8847,8229,9978,8660,7653,3296,3743,3310,3089,3147,3209,3256,4120,3766,2306,3515,2515,2404,3139,3321,2463,2185,3194,2812,3276,3237,1986,3137,2914,2615,2541,2600,2745,2915,2851,3886,3201,1753,3285,4405,2392,3272,3329,2355,3777,3398,2786,2645,2712,2449,3525,4103,2755,3897,2280,3794,4151,3797,3590,3266,4057,3378,2979,3295,3041,4076,2961,3517,4029,3488,2045,3210,2728,2768,3791,2161,3105,3148,3023,3636,2549,2855,4020,3846,3206,2963,2671,3081,3109,3725,3139,3146,4152,2028,3546,2615,3732],"x":["loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21"],"type":"box","marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"line":{"color":"rgba(31,119,180,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>plot_ly(data, y = ~resp, x = ~gen, type = &quot;box&quot;, width = 800, height = 480)</code></pre>
<div id="htmlwidget-31a3defda428c3790bb0" style="width:800px;height:480px;" class="plotly html-widget "></div>
<script type="application/json" data-for="htmlwidget-31a3defda428c3790bb0">{"x":{"visdat":{"36b476305216":["function () ","plotlyVisDat"]},"cur_data":"36b476305216","attrs":{"36b476305216":{"y":{},"x":{},"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"box"}},"layout":{"width":800,"height":480,"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":"gen","type":"category","categoryorder":"array","categoryarray":["geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23"]},"yaxis":{"domain":[0,1],"automargin":true,"title":"resp"},"hovermode":"closest","showlegend":false},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"fillcolor":"rgba(31,119,180,0.5)","y":[1275,1295,1620,1663,1015,1450,1465,1943,1285,950,1283,1168,1380,1413,1273,1273,1525,1078,1575,1553,1643,1225,910,1648,1740,1330,1800,1515,1250,1205,1218,1310,1165,1193,1498,1208,1360,1300,1438,1253,1573,1795,1260,1195,1505,1018,3600,2200,3500,3000,4000,3100,5200,6400,5000,6200,4200,6200,4000,3600,4600,3400,4000,4300,4000,4600,4000,5500,2800,5000,5000,2500,2800,3200,2000,2200,2800,3200,2200,2000,2400,1200,1800,2200,2000,2050,1600,2000,1000,900,1000,4200,2091,812,2254,1000,2045,1600,1094,1716,1670,1666,834,2355,1940,1973,1630,969,1094,1184,1853,1557,1831,1591,1125,825,2011,1650,1894,1570,1013,1296,1285,1660,1263,799,1541,1708,1577,1928,1908,1714,1410,1698,939,1913,1850,2128,6860,8693,8863,8620,8340,8617,9217,6973,8417,8070,8517,6363,6270,9410,8053,8043,8220,6360,6793,6637,7003,6513,6140,9370,9423,10003,9503,6013,6600,8983,10097,8417,6490,9530,5883,8130,10217,6943,7200,6193,9647,8803,9893,10440,5687,9747,2890,2372,2503,2162,2997,2395,2188,2712,1688,1752,1752,1437,1775,2743,2455,2373,2793,2548,2082,1967,2052,968,1840,2602,1485,1085,2342,2448,2453,1960,1857,2967,2252,2323,3157,2502,1900,1957,2280,1530,1337,2220,2372,3258,2595,2025,3167,4500,4333,4667,5333,4083,4167,5333,4833,4667,4000,4833,5667,5333,4000,3833,4167,4500,4000,5833,5800,4667,4333,3333,4167,4167,4833,5500,4333,4333,5333,5167,4833,3833,5000,5167,5000,3833,4133,4500,4333,3667,5500,5667,4667,4167,9560,5208,7176,6134,8681,7292,6319,7523,5833,7917,5903,7639,7940,7870,8218,7593,8912,6296,7407,5856,6296,5347,7616,8611,8171,7176,9144,7222,7940,8403,6713,8056,6134,6296,8264,7917,6759,6759,7685,6759,7361,8403,5139,6366,7292,7986,7158,7172,6806,6658,8072,7853,7744,8106,7011,8439,8403,8708,6956,8878,9242,7128,6550,8261,7903,7306,6164,8172,7592,7231,7792,8150,8675,9717,9236,8828,7522,6886,6089,8283,8556,7175,6764,7553,7761,7761,7789,7439,7694,9389,9219,8611,980,1020,840,910,990,720,820,870,1120,1090,1090,900,910,900,1020,1100,1080,1110,1080,1110,1140,990,920,1110,960,1000,1040,1100,1040,770,850,1120,1110,1100,1010,990,990,920,850,1210,960,910,1040,900,950,940,1542,1525,1692,1283,1217,2025,1875,1575,1725,3283,3025,2517,2817,3158,3150,2200,2267,2142,1683,1992,1642,2075,3025,3200,1950,2083,2025,1917,1850,1325,1292,1700,1450,2767,3117,3317,2158,2067,2475,1292,1417,1558,1858,1608,1417,1317,1008,1324,1488,1273,1086,1831,1416,1629,1116,1537,1884,898,1733,1429,1708,1429,1363,1473,1351,1618,1543,2096,1329,1025,1335,1492,1273,1635,1237,1680,1659,910,1069,1878,910,1512,1543,1702,904,1657,1653,1688,1600,1520,1514,1329,1316,1224,2026,2053,2132,2158,2211,3342,1737,1737,1553,1982,2368,1955,1729,1316,2024,1853,1645,1758,2092,2139,1626,2737,2368,1921,2474,2500,2184,1724,2539,2263,2026,1868,2684,1579,1579,1789,2974,2158,2053,2579,1842,1868,1618,1474,5798,3887,5899,4869,6846,6126,7769,6299,6019,6351,5357,6280,5661,5637,6369,5417,5893,6315,6315,5667,5720,4815,6280,5637,7101,5292,6839,5821,5952,6202,5804,6738,5512,4762,5964,6089,5893,5911,7714,5869,5530,5625,4429,5071,5780,5185,3992,3950,4208,3217,2500,2600,3100,3850,2675,3167,3817,2958,3750,3238,5033,3792,3683,3425,3092,3333,2250,4117,2083,3642,3350,2250,2825,3183,2033,2625,2908,3042,2850,5475,5583,4483,4783,4167,3558,2317,3505,3525,4133,3550,2575,4158,2706,1740,3060,1540,1980,2300,3120,2840,2840,3180,2520,2460,2400,3660,4340,4580,3220,2800,2860,1420,1900,2660,2120,2737,2420,3080,1840,2420,2420,2580,4400,3780,3560,3040,1860,2220,3020,2920,2560,3100,3140,3080,1400,1840,2560,2980,10167,8595,10690,7833,10500,8500,10762,9000,11190,8976,9500,11190,9905,11048,10833,8952,10214,10167,9976,10786,8738,9571,9262,11381,7333,10262,10524,9905,8714,9571,6524,9405,9048,9310,11167,10286,10167,10214,10595,10262,9810,9762,8857,8143,9976,10071,3390,3133,3508,3757,3621,2623,3624,3380,2627,3643,2751,3320,2942,3306,3807,3885,3138,3457,3355,4176,3767,3458,2552,3081,3919,3005,3762,3188,3661,2867,3780,3444,3636,3146,3409,2571,3229,3393,4310,4336,3244,4163,3097,3743,3617,2588,3900,2008,2375,3258,2208,2025,2500,2950,2242,2325,2667,2958,2675,2500,1833,2883,3617,3608,2733,2167,3508,3003,2742,2892,2958,3025,2500,3100,3408,1667,3692,3383,3600,2667,1750,2567,2317,2495,2792,2408,2767,2908,2008,3900,3392,2758,9778,8629,9185,7814,9203,10295,8710,9047,9973,10175,9985,9923,8438,9124,9560,9860,9195,9161,9073,8998,8580,8285,9396,8795,9023,7960,10083,9625,8995,9763,7214,8961,8378,8960,9503,9196,9473,7893,8300,8055,8924,8847,8229,9978,8660,7653,3296,3743,3310,3089,3147,3209,3256,4120,3766,2306,3515,2515,2404,3139,3321,2463,2185,3194,2812,3276,3237,1986,3137,2914,2615,2541,2600,2745,2915,2851,3886,3201,1753,3285,4405,2392,3272,3329,2355,3777,3398,2786,2645,2712,2449,3525,4103,2755,3897,2280,3794,4151,3797,3590,3266,4057,3378,2979,3295,3041,4076,2961,3517,4029,3488,2045,3210,2728,2768,3791,2161,3105,3148,3023,3636,2549,2855,4020,3846,3206,2963,2671,3081,3109,3725,3139,3146,4152,2028,3546,2615,3732],"x":["geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13"],"type":"box","marker":{"color":"rgba(31,119,180,1)","line":{"color":"rgba(31,119,180,1)"}},"line":{"color":"rgba(31,119,180,1)"},"xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<div id="summary-table-of-data" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Summary Table of
Data</h2>
<ul>
<li><code>Environment</code>: Environment/location name.</li>
<li><code>Type</code>: Experiment design.</li>
<li><code>nRep</code>: Number of replications.</li>
<li><code>nTotalEntries</code>: Total Number of entries
(genotypes).</li>
<li><code>nCheckEntries</code>: Total Number of check entries.</li>
<li><code>nTestEntries</code>: Total Number of test entries.</li>
<li><code>nonNA_nVar</code>: Total Number of non-missing entries.</li>
<li><code>nonNA_obs</code>: Total Number of non-missing
observations.</li>
<li><code>nonNA_AvRep</code>: Average Number of replications for
non-missing entries.</li>
<li><code>grandMean</code>: Grand mean mu.</li>
<li><code>checksMean</code>: Checks mean mu(c).</li>
<li><code>newEntriesMean</code>: New entries mean mu(L).</li>
<li><code>Trait</code>: Response trait name.</li>
<li><code>NumMissing</code>: Number of missing values.</li>
<li><code>Min</code>: Minimum value of response.</li>
<li><code>Max</code>: Maximum value of response.</li>
<li><code>Range</code>: Range of response values.</li>
<li><code>Median</code>: Median response value.</li>
<li><code>LowerQuartile</code>: Lower Quartile value (25%).</li>
<li><code>UpperQuartile</code>: Upper Quartile value (75%).</li>
<li><code>MeanRep</code>: Average number of replications.</li>
<li><code>MinRep</code>: Minimum number of replications.</li>
<li><code>MaxRep</code>: Maximum number of replications.</li>
</ul>
<pre class="r"><code>create_dt(summS[,c(1:12, 19:20, 24:32)])</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-89c70b1080806e243e32" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-89c70b1080806e243e32">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"],["loc_01","loc_02","loc_03","loc_04","loc_05","loc_06","loc_07","loc_08","loc_09","loc_10","loc_11","loc_12","loc_13","loc_14","loc_15","loc_16","loc_17","loc_18","loc_19","loc_20","loc_21"],["ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA","ALPHA"],[2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],[23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23],[4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4],[19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19],[23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23],[46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46,46],[2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],[1370.37,3318.48,1553.57,8047.91,2203.28,4598.15,7284.61,7834.83,990.87,2056.41,1441.02,2016.89,5876.28,3442.39,2721.8,9731.35,3400.2,2774.76,9018.59,3016.89,3270.7],[1295,2681.25,1079.62,8013.25,2142.75,4541.5,6336.75,7872.88,1088.75,2063.62,1578.38,1737.88,5167.5,3606.25,2527.5,9372,3523.12,2527,8963.88,3202.38,2893.5],[1386.24,3452.63,1653.34,8055.21,2216.03,4610.08,7484.16,7826.82,970.263,2054.89,1412.11,2075.63,6025.5,3407.89,2762.71,9807,3374.32,2826.92,9030.11,2977.84,3350.11],["GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha","GY_Calc_kgha"],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[910,900,799,5687,968,3167,5139,6089,720,1217,898,1224,3887,2033,1400,6524,2552,1667,7214,1753,2028],[1943,6400,2355,10440,3258,5833,9560,9717,1210,3317,2096,3342,7769,5583,4580,11381,4336,3900,10295,4405,4152],[1033,5500,1556,4753,2290,2666,4421,3628,490,2100,1198,2118,3882,3550,3180,4857,1784,2233,3081,2652,2124],[1305,3200,1640,8280,2266,4500,7384,7775,990,1933.5,1490,2003,5881,3387.5,2721.5,9940.5,3401,2750,9035,3138,3208],[1219.75,2200,1203.75,6676,1867.75,4167,6330.75,7172.75,910,1562.25,1285.75,1731,5553.75,2864.5,2325,9012,3134.25,2383.25,8592.25,2603.75,2961.5],[1522.5,4200,1883.75,9331.75,2502.75,5125.25,7974.5,8430,1090,2423,1648.5,2204.25,6280,3925,3080,10446.5,3722.5,3081.25,9545.75,3306.5,3776.25],[1370.37,3318.48,1553.57,8047.91,2203.28,4598.15,7284.61,7834.83,990.87,2056.41,1441.02,2016.89,5876.28,3442.39,2721.8,9731.35,3400.2,2774.76,9018.59,3016.89,3270.7],[1359.13,2402.17,1546.96,7695.3,2193.22,4585.48,7153.74,7664.43,987.391,1963.48,1422.83,1912,5857.39,3383.91,2706.35,9621.17,3357.39,2725.43,8802.96,2971.78,3184.65],[1381.61,4234.78,1560.17,8400.52,2213.35,4610.83,7415.48,8005.22,994.348,2149.35,1459.22,2121.78,5895.17,3500.87,2737.26,9841.52,3443,2824.09,9234.22,3062,3356.74]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Environment<\/th>\n      <th>Type<\/th>\n      <th>nRep<\/th>\n      <th>nTotalEntries<\/th>\n      <th>nCheckEntries<\/th>\n      <th>nTestEntries<\/th>\n      <th>nonNA_nVar<\/th>\n      <th>nonNA_obs<\/th>\n      <th>nonNA_AvRep<\/th>\n      <th>grandMean<\/th>\n      <th>checksMean<\/th>\n      <th>newLinesMean<\/th>\n      <th>Trait<\/th>\n      <th>NumMissing<\/th>\n      <th>Min<\/th>\n      <th>Max<\/th>\n      <th>Range<\/th>\n      <th>Median<\/th>\n      <th>LowerQuartile<\/th>\n      <th>UpperQuartile<\/th>\n      <th>MeanRep<\/th>\n      <th>MinRep<\/th>\n      <th>MaxRep<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8,9,10,11,12,14,15,16,17,18,19,20,21,22,23]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="detecting-outliers" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Detecting
Outliers</h2>
<p>These are the residuals and variograms from each trial available to
identify potential outliers (scres correspond to the standardized
conditional residuals). The interactive graphs are presented below for
exploration. However, in addition, some outliers are reported in the
database based on the pre-specified thresholds (see table below). Note
that some are identified as such (outlier = 1), and counterpart
observations (outlier = 0) for the same germplasms in other
replications. Close exploration of this database should ensure that the
corresponding observations are eliminated.</p>
<pre class="r"><code>select.environment &lt;- summS[!is.na(summS$sel_model) &amp; !is.na(summS$residualVar), &quot;Environment&quot;]
select.residual    &lt;- paste0(&quot;plots/residual_&quot;, select.environment, &quot;_&quot;, trait.name, &quot;.png&quot;)

names(select.residual) &lt;- select.environment

bsselect(select.residual, 
         type = &quot;iframe&quot;, 
         live_search = TRUE, 
         show_tick = TRUE, 
         elementId = &#39;residuals&#39;)$prepend[[1]]$html</code></pre>
<html>
<select id="HlEYEEQe6H" class="selectpicker" data-dropdown-align-right="false" data-dropup-auto="true" data-header="false" data-live-search="true" data-live-search-style="contains" data-show-tick="true" data-width="false" data-size="auto"><option value="plots/residual_loc_01_GY_Calc_kgha.png">loc_01</option>
<option value="plots/residual_loc_02_GY_Calc_kgha.png">loc_02</option>
<option value="plots/residual_loc_03_GY_Calc_kgha.png">loc_03</option>
<option value="plots/residual_loc_04_GY_Calc_kgha.png">loc_04</option>
<option value="plots/residual_loc_05_GY_Calc_kgha.png">loc_05</option>
<option value="plots/residual_loc_06_GY_Calc_kgha.png">loc_06</option>
<option value="plots/residual_loc_07_GY_Calc_kgha.png">loc_07</option>
<option value="plots/residual_loc_08_GY_Calc_kgha.png">loc_08</option>
<option value="plots/residual_loc_09_GY_Calc_kgha.png">loc_09</option>
<option value="plots/residual_loc_10_GY_Calc_kgha.png">loc_10</option>
<option value="plots/residual_loc_11_GY_Calc_kgha.png">loc_11</option>
<option value="plots/residual_loc_12_GY_Calc_kgha.png">loc_12</option>
<option value="plots/residual_loc_13_GY_Calc_kgha.png">loc_13</option>
<option value="plots/residual_loc_14_GY_Calc_kgha.png">loc_14</option>
<option value="plots/residual_loc_15_GY_Calc_kgha.png">loc_15</option>
<option value="plots/residual_loc_16_GY_Calc_kgha.png">loc_16</option>
<option value="plots/residual_loc_17_GY_Calc_kgha.png">loc_17</option>
<option value="plots/residual_loc_18_GY_Calc_kgha.png">loc_18</option>
<option value="plots/residual_loc_19_GY_Calc_kgha.png">loc_19</option>
<option value="plots/residual_loc_20_GY_Calc_kgha.png">loc_20</option>
<option value="plots/residual_loc_21_GY_Calc_kgha.png">loc_21</option></select>
<iframe src="plots/residual_loc_01_GY_Calc_kgha.png" frameborder="0" height="500" width="100%" id="VJx6ToOQFo"></iframe>
<script>$(document).ready(function(){
                   $("#HlEYEEQe6H").change(function(){
                   $("#VJx6ToOQFo").attr("src",$(this).val());

                   });

    });</script>
</html>
<pre class="r"><code>select.environment &lt;- summS[!is.na(summS$sel_model) &amp; !is.na(summS$residualVar), &quot;Environment&quot;]
select.variogram   &lt;- paste0(&quot;plots/variogram_&quot;, select.environment, &quot;_&quot;, trait.name, &quot;.png&quot;)

names(select.variogram) &lt;- select.environment

bsselect(select.variogram, 
         type = &quot;iframe&quot;, 
         live_search = TRUE, 
         show_tick = TRUE, 
         elementId = &quot;variograms&quot;)$prepend[[1]]$html</code></pre>
<html>
<select id="5vUyla4Bqo" class="selectpicker" data-dropdown-align-right="false" data-dropup-auto="true" data-header="false" data-live-search="true" data-live-search-style="contains" data-show-tick="true" data-width="false" data-size="auto"><option value="plots/variogram_loc_01_GY_Calc_kgha.png">loc_01</option>
<option value="plots/variogram_loc_02_GY_Calc_kgha.png">loc_02</option>
<option value="plots/variogram_loc_03_GY_Calc_kgha.png">loc_03</option>
<option value="plots/variogram_loc_04_GY_Calc_kgha.png">loc_04</option>
<option value="plots/variogram_loc_05_GY_Calc_kgha.png">loc_05</option>
<option value="plots/variogram_loc_06_GY_Calc_kgha.png">loc_06</option>
<option value="plots/variogram_loc_07_GY_Calc_kgha.png">loc_07</option>
<option value="plots/variogram_loc_08_GY_Calc_kgha.png">loc_08</option>
<option value="plots/variogram_loc_09_GY_Calc_kgha.png">loc_09</option>
<option value="plots/variogram_loc_10_GY_Calc_kgha.png">loc_10</option>
<option value="plots/variogram_loc_11_GY_Calc_kgha.png">loc_11</option>
<option value="plots/variogram_loc_12_GY_Calc_kgha.png">loc_12</option>
<option value="plots/variogram_loc_13_GY_Calc_kgha.png">loc_13</option>
<option value="plots/variogram_loc_14_GY_Calc_kgha.png">loc_14</option>
<option value="plots/variogram_loc_15_GY_Calc_kgha.png">loc_15</option>
<option value="plots/variogram_loc_16_GY_Calc_kgha.png">loc_16</option>
<option value="plots/variogram_loc_17_GY_Calc_kgha.png">loc_17</option>
<option value="plots/variogram_loc_18_GY_Calc_kgha.png">loc_18</option>
<option value="plots/variogram_loc_19_GY_Calc_kgha.png">loc_19</option>
<option value="plots/variogram_loc_20_GY_Calc_kgha.png">loc_20</option>
<option value="plots/variogram_loc_21_GY_Calc_kgha.png">loc_21</option></select>
<iframe src="plots/variogram_loc_01_GY_Calc_kgha.png" frameborder="0" height="500" width="100%" id="jQrSZESOmV"></iframe>
<script>$(document).ready(function(){
                   $("#5vUyla4Bqo").change(function(){
                   $("#jQrSZESOmV").attr("src",$(this).val());

                   });

    });</script>
</html>
<pre class="r"><code>select.environment &lt;- summS[!is.na(summS$sel_model) &amp; trial.type$has.spatial != &quot;NO&quot; &amp; !is.na(summS$residualVar), &quot;Environment&quot;]
select.residual    &lt;- paste0(&quot;plots/res_layout_&quot;, select.environment, &quot;_&quot;, trait.name, &quot;.png&quot;)

names(select.residual) &lt;- select.environment

if (length(select.environment) != 0) {
  bsselect(select.residual, 
           type = &quot;iframe&quot;, 
           live_search = TRUE, 
           show_tick = TRUE, 
           elementId = &#39;res_layouts&#39;)$prepend[[1]]$html
}</code></pre>
<html>
<select id="O7MUIRcBau" class="selectpicker" data-dropdown-align-right="false" data-dropup-auto="true" data-header="false" data-live-search="true" data-live-search-style="contains" data-show-tick="true" data-width="false" data-size="auto"><option value="plots/res_layout_loc_01_GY_Calc_kgha.png">loc_01</option>
<option value="plots/res_layout_loc_02_GY_Calc_kgha.png">loc_02</option>
<option value="plots/res_layout_loc_03_GY_Calc_kgha.png">loc_03</option>
<option value="plots/res_layout_loc_04_GY_Calc_kgha.png">loc_04</option>
<option value="plots/res_layout_loc_05_GY_Calc_kgha.png">loc_05</option>
<option value="plots/res_layout_loc_06_GY_Calc_kgha.png">loc_06</option>
<option value="plots/res_layout_loc_07_GY_Calc_kgha.png">loc_07</option>
<option value="plots/res_layout_loc_08_GY_Calc_kgha.png">loc_08</option>
<option value="plots/res_layout_loc_09_GY_Calc_kgha.png">loc_09</option>
<option value="plots/res_layout_loc_10_GY_Calc_kgha.png">loc_10</option>
<option value="plots/res_layout_loc_11_GY_Calc_kgha.png">loc_11</option>
<option value="plots/res_layout_loc_12_GY_Calc_kgha.png">loc_12</option>
<option value="plots/res_layout_loc_13_GY_Calc_kgha.png">loc_13</option>
<option value="plots/res_layout_loc_14_GY_Calc_kgha.png">loc_14</option>
<option value="plots/res_layout_loc_15_GY_Calc_kgha.png">loc_15</option>
<option value="plots/res_layout_loc_16_GY_Calc_kgha.png">loc_16</option>
<option value="plots/res_layout_loc_17_GY_Calc_kgha.png">loc_17</option>
<option value="plots/res_layout_loc_18_GY_Calc_kgha.png">loc_18</option>
<option value="plots/res_layout_loc_19_GY_Calc_kgha.png">loc_19</option>
<option value="plots/res_layout_loc_20_GY_Calc_kgha.png">loc_20</option>
<option value="plots/res_layout_loc_21_GY_Calc_kgha.png">loc_21</option></select>
<iframe src="plots/res_layout_loc_01_GY_Calc_kgha.png" frameborder="0" height="500" width="100%" id="UmPLKglCgY"></iframe>
<script>$(document).ready(function(){
                   $("#O7MUIRcBau").change(function(){
                   $("#UmPLKglCgY").attr("src",$(this).val());

                   });

    });</script>
</html>
<pre class="r"><code>create_dt(temp)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-45a79d950237a1c47ec9" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-45a79d950237a1c47ec9">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480","481","482","483","484","485","486","487","488","489","490","491","492","493","494","495","496","497","498","499","500","501","502","503","504","505","506","507","508","509","510","511","512","513","514","515","516","517","518","519","520","521","522","523","524","525","526","527","528","529","530","531","532","533","534","535","536","537","538","539","540","541","542","543","544","545","546","547","548","549","550","551","552","553","554","555","556","557","558","559","560","561","562","563","564","565","566","567","568","569","570","571","572","573","574","575","576","577","578","579","580","581","582","583","584","585","586","587","588","589","590","591","592","593","594","595","596","597","598","599","600","601","602","603","604","605","606","607","608","609","610","611","612","613","614","615","616","617","618","619","620","621","622","623","624","625","626","627","628","629","630","631","632","633","634","635","636","637","638","639","640","641","642","643","644","645","646","647","648","649","650","651","652","653","654","655","656","657","658","659","660","661","662","663","664","665","666","667","668","669","670","671","672","673","674","675","676","677","678","679","680","681","682","683","684","685","686","687","688","689","690","691","692","693","694","695","696","697","698","699","700","701","702","703","704","705","706","707","708","709","710","711","712","713","714","715","716","717","718","719","720","721","722","723","724","725","726","727","728","729","730","731","732","733","734","735","736","737","738","739","740","741","742","743","744","745","746","747","748","749","750","751","752","753","754","755","756","757","758","759","760","761","762","763","764","765","766","767","768","769","770","771","772","773","774","775","776","777","778","779","780","781","782","783","784","785","786","787","788","789","790","791","792","793","794","795","796","797","798","799","800","801","802","803","804","805","806","807","808","809","810","811","812","813","814","815","816","817","818","819","820","821","822","823","824","825","826","827","828","829","830","831","832","833","834","835","836","837","838","839","840","841","842","843","844","845","846","847","848","849","850","851","852","853","854","855","856","857","858","859","860","861","862","863","864","865","866","867","868","869","870","871","872","873","874","875","876","877","878","879","880","881","882","883","884","885","886","887","888","889","890","891","892","893","894","895","896","897","898","899","900","901","902","903","904","905","906","907","908","909","910","911","912","913","914","915","916","917","918","919","920","921","922","923","924","925","926","927","928","929","930","931","932","933","934","935","936","937","938","939","940","941","942","943","944","945","946","947","948","949","950","951","952","953","954","955","956","957","958","959","960","961","962","963","964","965","966"],[1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4],[1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,11,1,2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9,10,12],[9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,18,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,8,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16],["loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21"],["geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_17","geno_12","geno_22","geno_05","geno_10","geno_16","geno_06","geno_04","geno_18","geno_21","geno_11","geno_15","geno_23","geno_09","geno_07","geno_03","geno_20","geno_14","geno_08","geno_02","geno_01","geno_19","geno_13"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],[1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6],[1275,1295,1620,1663,1015,1450,1465,1943,1285,950,1283,1168,1380,1413,1273,1273,1525,1078,1575,1553,1643,1225,910,1648,1740,1330,1800,1515,1250,1205,1218,1310,1165,1193,1498,1208,1360,1300,1438,1253,1573,1795,1260,1195,1505,1018,3600,2200,3500,3000,4000,3100,5200,6400,5000,6200,4200,6200,4000,3600,4600,3400,4000,4300,4000,4600,4000,5500,2800,5000,5000,2500,2800,3200,2000,2200,2800,3200,2200,2000,2400,1200,1800,2200,2000,2050,1600,2000,1000,900,1000,4200,2091,812,2254,1000,2045,1600,1094,1716,1670,1666,834,2355,1940,1973,1630,969,1094,1184,1853,1557,1831,1591,1125,825,2011,1650,1894,1570,1013,1296,1285,1660,1263,799,1541,1708,1577,1928,1908,1714,1410,1698,939,1913,1850,2128,6860,8693,8863,8620,8340,8617,9217,6973,8417,8070,8517,6363,6270,9410,8053,8043,8220,6360,6793,6637,7003,6513,6140,9370,9423,10003,9503,6013,6600,8983,10097,8417,6490,9530,5883,8130,10217,6943,7200,6193,9647,8803,9893,10440,5687,9747,2890,2372,2503,2162,2997,2395,2188,2712,1688,1752,1752,1437,1775,null,2455,2373,2793,2548,2082,1967,2052,968,1840,2602,1485,1085,2342,2448,2453,1960,1857,2967,2252,2323,3157,2502,1900,1957,2280,1530,1337,2220,2372,3258,2595,2025,3167,4500,4333,4667,5333,4083,4167,5333,4833,4667,4000,4833,5667,5333,4000,3833,4167,4500,4000,5833,5800,4667,4333,3333,4167,4167,4833,5500,4333,4333,5333,5167,4833,3833,5000,5167,5000,3833,4133,4500,4333,3667,5500,5667,4667,4167,null,5208,7176,6134,8681,7292,6319,7523,5833,7917,5903,7639,7940,7870,8218,7593,8912,6296,7407,5856,6296,5347,7616,8611,8171,7176,9144,7222,7940,8403,6713,8056,6134,6296,8264,7917,6759,6759,7685,6759,7361,8403,5139,6366,7292,7986,7158,7172,6806,6658,8072,7853,7744,8106,7011,8439,8403,8708,6956,8878,9242,7128,6550,8261,7903,7306,6164,8172,7592,7231,7792,8150,8675,9717,9236,8828,7522,6886,6089,8283,8556,7175,6764,7553,7761,7761,7789,7439,7694,9389,9219,8611,980,1020,840,910,990,720,820,870,1120,1090,1090,900,910,900,1020,1100,1080,1110,1080,1110,1140,990,920,1110,960,1000,1040,1100,1040,770,850,1120,1110,1100,1010,990,990,920,850,1210,960,910,1040,900,950,940,1542,1525,1692,1283,1217,2025,1875,1575,1725,3283,3025,2517,2817,3158,3150,2200,2267,2142,1683,1992,1642,2075,3025,3200,1950,2083,2025,1917,1850,1325,1292,1700,1450,2767,3117,3317,2158,2067,2475,1292,1417,1558,1858,1608,1417,1317,1008,1324,1488,1273,1086,1831,1416,1629,1116,1537,1884,898,1733,1429,1708,1429,1363,1473,1351,1618,1543,2096,1329,1025,1335,1492,1273,1635,1237,1680,1659,910,1069,1878,910,1512,1543,1702,904,1657,1653,1688,1600,1520,1514,1329,1316,1224,2026,2053,2132,2158,2211,3342,1737,1737,1553,1982,2368,1955,1729,1316,2024,1853,1645,1758,2092,2139,1626,2737,2368,1921,2474,2500,2184,1724,2539,2263,2026,1868,2684,1579,1579,1789,2974,2158,2053,2579,1842,1868,1618,1474,5798,3887,5899,4869,6846,6126,7769,6299,6019,6351,5357,6280,5661,5637,6369,5417,5893,6315,6315,5667,5720,4815,6280,5637,7101,5292,6839,5821,5952,6202,5804,6738,5512,4762,5964,6089,5893,5911,7714,5869,5530,5625,4429,5071,5780,5185,3992,3950,4208,3217,2500,2600,3100,3850,2675,3167,3817,2958,3750,3238,5033,3792,3683,3425,3092,3333,2250,4117,2083,3642,3350,2250,2825,3183,2033,2625,2908,3042,2850,5475,5583,4483,4783,4167,3558,2317,3505,3525,4133,3550,2575,4158,2706,1740,3060,1540,1980,2300,3120,2840,2840,3180,2520,2460,2400,3660,4340,4580,3220,2800,2860,1420,1900,2660,2120,2737,2420,3080,1840,2420,2420,2580,4400,3780,3560,3040,1860,2220,3020,2920,2560,3100,3140,3080,1400,1840,2560,2980,10167,8595,10690,7833,10500,8500,10762,9000,11190,8976,9500,11190,9905,11048,10833,8952,10214,10167,9976,10786,8738,9571,9262,11381,7333,10262,10524,9905,8714,9571,6524,9405,9048,9310,11167,10286,10167,10214,10595,10262,9810,9762,8857,8143,9976,10071,3390,3133,3508,3757,3621,2623,3624,3380,2627,3643,2751,3320,2942,3306,3807,3885,3138,3457,3355,4176,3767,3458,2552,3081,3919,3005,3762,3188,3661,2867,3780,3444,3636,3146,3409,2571,3229,3393,4310,4336,3244,4163,3097,3743,3617,2588,3900,2008,2375,3258,2208,2025,2500,2950,2242,2325,2667,2958,2675,2500,1833,2883,3617,3608,2733,2167,3508,3003,2742,2892,2958,3025,2500,3100,3408,1667,3692,3383,3600,2667,1750,2567,2317,2495,2792,2408,2767,2908,2008,3900,3392,2758,9778,8629,9185,7814,9203,10295,8710,9047,9973,10175,9985,9923,8438,9124,9560,9860,9195,9161,9073,8998,8580,8285,9396,8795,9023,7960,null,9625,8995,9763,7214,8961,8378,8960,9503,9196,9473,7893,8300,8055,8924,8847,8229,9978,8660,7653,3296,3743,3310,3089,3147,3209,3256,4120,3766,2306,3515,2515,2404,3139,3321,2463,2185,3194,2812,3276,3237,1986,3137,2914,2615,2541,2600,2745,2915,2851,3886,3201,1753,3285,4405,2392,3272,3329,2355,3777,3398,2786,2645,2712,2449,3525,4103,2755,3897,2280,3794,4151,3797,3590,3266,4057,3378,2979,3295,3041,4076,2961,3517,4029,3488,2045,3210,2728,2768,3791,2161,3105,3148,3023,3636,2549,2855,4020,3846,3206,2963,2671,3081,3109,3725,3139,3146,4152,2028,3546,2615,3732],[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,638,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714,715,716,717,718,719,720,721,722,723,724,725,726,727,728,729,730,731,732,733,734,735,736,737,738,739,740,741,742,743,744,745,746,747,748,749,750,751,752,753,754,755,756,757,758,759,760,761,762,763,764,765,766,767,768,769,770,771,772,773,774,775,776,777,778,779,780,781,782,783,784,785,786,787,788,789,790,791,792,793,794,795,796,797,798,799,800,801,802,803,804,805,806,807,808,809,810,811,812,813,814,815,816,817,818,819,820,821,822,823,824,825,826,827,828,829,830,831,832,833,834,835,836,837,838,839,840,841,842,843,844,845,846,847,848,849,850,851,852,853,854,855,856,857,858,859,860,861,862,863,864,865,866,867,868,869,870,871,872,873,874,875,876,877,878,879,880,881,882,883,884,885,886,887,888,889,890,891,892,893,894,895,896,897,898,899,900,901,902,903,904,905,906,907,908,909,910,911,912,913,914,915,916,917,918,919,920,921,922,923,924,925,926,927,928,929,930,931,932,933,934,935,936,937,938,939,940,941,942,943,944,945,946,947,948,949,950,951,952,953,954,955,956,957,958,959,960,961,962,963,964,965,966],[0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,1,0,0,0],[-0.73055,-0.661957,0.638345,1.21572,-2.16877,0.209144,0.194974,2.60671,-0.139305,-1.64391,0.289934,-0.368565,-0.0315082,0.177082,-0.672292,-0.0480654,0.622622,-1.61568,0.907401,1.09474,1.60741,-0.323151,-1.31698,0.492936,1.26285,-1.16724,1.76024,0.320714,-0.952323,-0.524219,-0.803296,0.261809,-0.72394,-0.63919,1.32983,-0.816571,0.0397107,-0.67704,0.424653,-0.884979,0.620863,1.53312,-0.591404,-0.542514,0.966338,-0.659039,0.878493,-1.73999,0.632227,-0.788476,0.455226,-1.42613,0.548585,1.75871,-0.681798,1.32114,-1.29334,1.07974,-0.679023,-0.383864,1.16765,-0.393717,-0.772502,0.140365,0.241666,0.175246,-0.811814,2.60717,-2.12409,1.82366,1.33201,-2.0697,0.198772,0.132739,-0.521276,0.161713,0.330861,0.338962,-0.166122,0.191125,-0.655203,-0.217151,0.0370989,-0.0860885,-0.307464,0.123558,0.0239237,-0.581872,0.512104,-1.09977,-1.13667,2.24235,1.35122,-1.34949,-0.254281,-0.724122,0.280004,1.38879,-1.64047,-0.0352785,0.485245,0.229261,0.340671,1.36196,-0.256987,1.83618,0.0291004,-1.5352,-0.547643,-2.02912,1.01157,0.13336,0.51617,0.272676,-0.847473,-1.08099,-0.182393,-0.270584,0.825928,0.0638229,0.852836,-1.18791,0.409371,1.69013,-0.713876,-1.07618,-0.675203,0.516875,-0.440113,1.62652,1.2438,-0.0339508,-1.44044,0.450246,-0.0553319,-0.881497,-0.215958,0.737575,-0.67042,0.388663,0.743953,0.0897538,0.0631019,0.622982,1.55158,-0.709232,0.23626,0.938207,0.531238,-0.760122,-1.40855,0.790834,0.137622,0.182703,0.784004,-0.724219,-0.403378,-0.522169,-0.112779,-0.688141,-1.10713,0.732731,-0.0284372,1.25818,0.950804,-1.68886,-1.18043,0.198111,1.09604,-0.245292,-0.835098,1.37096,-2.24785,-0.234256,1.26015,-1.78332,-0.856098,-0.960793,1.17218,0.068213,0.340635,1.68671,-1.86483,1.79676,0.72728,-0.685662,0.365231,0.12406,1.07693,0.291011,-0.380938,0.442866,-0.0846315,-0.905406,0.153508,-0.46181,-1.32651,3.01252,-1.19059,-0.231558,0.682431,-0.124313,-0.352709,-0.194719,0.974299,-1.26016,-0.722414,0.719692,-0.833836,-0.911359,-0.112923,0.965972,0.639034,-0.689214,-0.88826,1.23312,-1.07788,-0.510764,2.09352,0.962922,-0.654232,-0.129945,0.0984044,-0.276498,-2.96905,0.45558,1.03155,0.86701,0.383797,-0.0936278,-2.32375,0.610704,-0.184479,0.0485802,1.54347,-1.0072,-0.49429,1.09892,-0.192527,0.407496,-0.967872,0.745795,1.81203,0.747188,-0.918926,-0.821289,-0.826183,0.573937,-1.43551,1.26208,1.15903,-0.589702,-0.243764,-1.92791,-0.0220597,-0.136645,0.40575,1.62814,-0.746834,-0.0137665,0.648196,0.113324,0.696369,-1.39433,1.28935,0.891029,0.690347,-1.17093,-0.383019,-0.0886071,0.178047,-1.80403,1.18125,1.15362,-0.461576,-0.882808,3.03226,-1.66654,-0.68303,-1.15224,0.578815,-0.693321,-0.540501,-0.451522,-0.949792,1.45548,-0.641712,0.691259,-0.201698,0.572341,0.459003,0.24581,1.9526,-1.46006,0.418398,-0.801469,0.282283,-1.35063,1.05668,-0.692145,-0.186594,0.283613,0.797654,-1.19856,0.190744,1.18018,0.157084,1.48669,-1.11744,-0.377843,0.620051,-0.47538,-0.148077,-0.301738,0.632025,0.0300071,-0.244799,1.29192,-0.63758,-2.37952,-0.0158948,1.02627,-0.621261,-0.0465567,-0.824847,0.260392,-0.295645,-0.063516,-0.0408749,0.779134,-0.834445,-0.123854,0.0429855,1.60516,-1.36824,1.90935,1.6168,-1.83352,-1.30857,1.83293,0.113444,-0.724912,-1.34504,0.822707,0.433372,-0.189423,-1.07267,-0.155834,1.14147,1.40563,1.72435,0.206513,-1.68514,-1.80566,-1.11421,1.02049,0.50424,-0.848653,-0.684858,-0.0914391,0.616833,0.54443,-0.881944,-0.713207,0.0798409,1.10718,0.45946,0.48102,0.73322,-0.160283,-0.755981,0.449872,-0.0285845,-1.27529,-1.51887,0.0181519,2.65909,-0.0527817,0.0723236,-0.931981,-1.44458,-0.923289,0.221398,1.30382,-0.0300536,0.0850634,2.01024,-0.977113,1.12054,0.361737,-1.08864,0.549439,0.481126,-0.367612,0.201906,0.690012,-0.791666,-0.0752535,-1.0906,0.625835,-0.292124,0.554008,-0.0532615,0.862425,-2.31791,0.876835,-0.146218,1.91377,0.482784,-0.580715,0.310352,-1.02039,-1.81565,1.11104,-0.272016,-0.68226,0.0260513,-0.471311,-1.41736,0.911485,0.486354,-0.751699,-1.86383,1.77338,0.527879,0.534033,0.962549,1.76624,0.0451103,-1.48484,-0.703633,0.895919,-0.761745,0.290788,-0.638903,-0.0926256,0.926443,2.1971,-0.999669,-0.405864,0.169638,-0.376213,1.18295,-1.13589,-0.483323,-1.3656,-1.09501,1.00441,1.97072,1.44054,0.502077,-0.907696,0.165024,-0.791256,-1.30396,0.294371,0.870346,0.151336,0.00994378,-1.12548,-1.38464,-0.681809,1.58062,-1.1028,-1.10095,1.11559,0.512241,0.372018,-0.975408,0.709303,1.04368,-1.36073,0.415559,-0.485649,1.13403,-0.675421,-0.787487,0.0942811,-0.559733,0.300338,0.0223894,1.52452,0.241951,-0.476399,0.399274,0.312391,-0.221544,0.793287,0.336228,1.07214,1.67526,-1.39497,-0.650818,0.882524,-1.69568,-0.320795,0.640998,-0.171593,-2.9173,-0.0458491,0.824532,0.803011,0.999488,0.258444,0.0148853,-1.35588,-0.692057,-0.725783,-0.23435,0.245788,-0.790513,0.769738,-0.423881,2.57546,-0.617255,-0.840276,-0.110416,0.226162,1.81301,0.163986,-0.915949,-1.09375,0.514531,0.18723,-0.345854,-0.206974,0.370207,1.35748,-1.10865,0.947432,0.19934,-1.86744,1.04684,0.855776,0.463489,-1.83108,0.970612,0.196483,-0.470019,-1.02015,1.69275,-0.262319,-0.594611,-0.775311,2.24581,-0.0515913,-0.214154,-0.122205,0.0905719,0.40936,-0.234907,-1.52869,1.29088,-2.06043,-1.11026,-0.817784,0.693584,-0.839582,2.78058,-0.299725,-0.347164,1.27776,-0.49706,-0.278288,-0.413289,0.391347,0.867918,-0.766858,0.449183,0.0455507,-0.217915,-0.206269,0.665458,-1.4752,1.04233,-0.930577,1.51427,-0.0165331,0.135994,-1.02994,0.431876,0.573199,0.0481109,1.37673,-0.91673,-0.936933,0.00206047,-0.0357151,0.0409428,-1.71774,2.92373,0.233664,-0.592958,0.091758,-0.260192,-1.54896,0.759043,-0.176352,0.617692,0.594177,1.04634,-0.607203,-0.906543,-0.613015,-0.145518,1.10126,-0.641174,-0.700281,0.909447,-0.195379,-0.170546,-0.985732,0.791852,0.957934,0.235342,0.304945,-0.270542,-0.0216637,-1.00398,1.96505,-2.26925,0.306129,0.10658,-1.94994,0.268388,0.880219,-1.85996,-0.254838,-0.138741,-1.08229,-0.596505,1.95541,1.96506,1.21016,1.10091,0.683592,-1.10813,-1.79336,0.787178,-0.33782,0.585374,0.0573375,-1.04321,0.302599,-0.056686,-1.58525,2.55575,-1.08261,-0.406762,-0.629332,1.30278,-0.980967,-0.399535,1.18651,-0.90275,0.190652,-1.25658,-0.213348,2.68213,0.814303,-0.414884,-0.636858,0.940279,-1.12574,-0.571367,1.34578,-1.13275,-0.87535,-0.209838,1.91802,-1.16205,0.23189,-0.328432,-0.292301,1.19987,-0.0512498,1.16513,0.326151,-1.54328,-1.10677,0.34061,0.697226,-1.54361,0.634631,0.130994,0.731592,-0.855241,-0.484751,0.656416,-0.122642,0.966365,-0.47254,-0.136134,-0.43539,-0.635217,-1.02067,0.00329586,-0.519676,0.471815,-1.13878,-0.040742,2.46082,0.151606,1.4177,0.473183,-0.798469,-0.0471149,0.914995,0.638401,0.578895,-0.79834,-0.891488,-1.06252,0.824915,-2.68064,0.509694,1.72108,0.903402,-0.0248378,-0.105075,-2.49954,-1.09108,-0.171902,0.049562,1.09891,1.14238,0.6742,0.747958,1.15573,0.371831,-0.365218,0.302517,-0.598507,-1.79775,-0.351761,0.297954,0.347268,0.647297,-1.20612,-0.550367,1.60495,0.38965,1.73627,-1.25066,-0.510522,2.11976,-1.01781,-1.12609,0.496846,-0.625396,0.925688,0.796282,-0.564752,-0.356242,-1.52077,-0.168905,0.895854,-0.842384,-0.364004,0.313131,1.49379,0.679494,-1.33911,-2.03877,-0.496822,-0.88551,1.07301,0.359828,-0.651216,0.703761,-0.414365,-0.776142,-0.303609,-1.55372,1.71218,1.13403,0.324329,1.71696,-0.971206,-0.220352,1.39299,-1.17475,1.17675,0.241521,-0.882103,-0.790201,-1.5053,0.570954,0.65506,-0.0194078,-0.438911,-1.51855,0.353387,0.243995,0.00970983,-0.067523,0.394308,-1.12845,2.32753,0.7988,-1.64603,-0.125898,0.799121,0.26294,0.315278,-1.91588,0.00977438,-0.00147519,1.00368,1.3353,1.33876,-1.44435,1.51219,-0.2886,-0.131754,-0.419815,-1.20729,-0.32169,0.0341053,-0.837243,0.717007,-0.43694,-0.123218,0.241944,-0.960895,-0.125411,1.9245,0.0349251,-1.0207,-0.346797,0.893769,-0.324617,-2.66419,0.700685,0.533762,-0.555027,0.485117,0.680813,1.55157,1.26994,0.341631,-0.518944,-0.603563,1.16969,-0.0815438,-0.501603,-0.112451,0.923329,-0.769841,-0.659554,-0.391472,0.0594357,-0.844021,-0.180336,3.24944,0.145322,-0.786421,0.246362,-1.08561,0.541356,0.264094,-1.12612,1.08431,0.651488,0.175119,-1.20588,-1.15254,-1.38455,0.524017,0.48793,-0.205814,1.82621,-0.0300977,-1.25369,0.215969,1.10968,0.407026,-0.600348,0.0963708,0.100697,-0.144745,1.77269,0.865997,-1.57906,0.488608,-1.11989,-1.25627,0.21063,0.218709,-0.755246,-1.29268,0.527437,0.0167656,0.300333,1.27746,-1.50595,0.665779,-0.0208646,-0.598207,-0.649502,-0.914765,-0.254241,-0.146669,-0.462677,1.57262,0.0695788,-2.74553,0.228479,2.33554,-0.965773,0.584791,0.929941,-0.979491,1.62578,0.961761,-0.518579,-0.617783,-0.282296,-0.708228,1.54056,0.811276,-0.603439,0.129077,-1.8379,0.52065,2.03102,1.25798,-1.31508,-0.321797,0.468732,-0.472582,0.050324,0.326826,-0.818109,2.0526,-0.99135,0.145189,1.78664,0.59344,-2.30677,-1.05696,-0.211741,-0.385503,0.513725,-1.49422,0.066277,-0.318337,-0.0633561,1.11023,-1.28107,-0.312831,0.450812,1.38218,0.363593,-1.32024,-0.375852,-0.523548,-0.915095,0.676417,1.24334,1.07685,1.54915,-0.927291,-0.350461,-0.86977,0.266707]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>row<\/th>\n      <th>col<\/th>\n      <th>TRIAL_INSTANCE<\/th>\n      <th>trial<\/th>\n      <th>gen<\/th>\n      <th>block<\/th>\n      <th>ibk<\/th>\n      <th>resp<\/th>\n      <th>id<\/th>\n      <th>check<\/th>\n      <th>scres<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[1,2,3,6,7,8,9,10,11]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<ul>
<li>Description of graphical/numerical step 1 output:
<ul>
<li><code>residual.plots</code>: Residual plots.</li>
<li><code>variogram.plots</code>: Variogram fitted model.</li>
</ul></li>
</ul>
<blockquote>
<p><em>Check your residual plots <a
href="./outputs/Study_GY_Calc_kgha_Stage-1_Residual_Plots.pdf">here</a></em></p>
</blockquote>
<blockquote>
<p><em>Check your variogram fitted model <a
href="./outputs/Study_GY_Calc_kgha_Stage-1_Variogram_Plots.pdf">here</a></em></p>
</blockquote>
<blockquote>
<p><em>Check your residual layouts <a
href="./outputs/Study_GY_Calc_kgha_Stage-1_Residual_Layouts.pdf">here</a></em></p>
</blockquote>
</div>
<div id="summary-table-of-final-models" class="section level2"
number="4.4">
<h2><span class="header-section-number">4.4</span> Summary Table of
Final Models</h2>
<p>The following table presents a summary of the final models selected
for each of the individual trials. The final model were fitted after the
elimination of outliers (outlier = 1). The procedure in this case fits a
total of 26 different spatial models (only 8 in case of Augmented design
trials) and selects the best model according to the specified
criteria.</p>
<p>It is critical to explore this output database, as from here some of
the trials with poor performance (e.g., heritVC &lt; 0.10) should be
eliminated as they might have installing or phenotypic issues and
therefore they are unlikely to contribute to the downstream MET
analyses. This summary table has several relevant statistics that can be
used in combination to assess their quality.</p>
<ul>
<li><code>Environment</code>: Environment/location name.</li>
<li><code>sel_model</code>: Selected model (see next section).</li>
<li><code>heritVC</code>: The heritability calculated based on variance
components.</li>
<li><code>heritPEV</code>: The heritability calculated based on the
predictor error variance (only for genotypes considered as
<em>random</em>).</li>
<li><code>genotypicVar</code>: Genotypic Variance.</li>
<li><code>residualVar</code>: Residual Variance.</li>
<li><code>CV</code>: Coefficient of Variation.</li>
<li><code>LSD</code>: Least Significant Difference.</li>
<li><code>Variance</code>: Total Variance of the response variable.</li>
<li><code>SD</code>: Response Standard Deviation.</li>
<li><code>WaldStatistic</code>: Wald Chi-Squared Test statistic (mixed
model ANOVA-like table).</li>
<li><code>WaldDF</code>: Wald-test Degree of Freedom.</li>
<li><code>Pvalue</code>: p-value for the test line.</li>
</ul>
<pre class="r"><code>create_dt(summS[,c(1, 13:18, 21:23, 33:35)])</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-53ad5756d249509facbc" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-53ad5756d249509facbc">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"],["loc_01","loc_02","loc_03","loc_04","loc_05","loc_06","loc_07","loc_08","loc_09","loc_10","loc_11","loc_12","loc_13","loc_14","loc_15","loc_16","loc_17","loc_18","loc_19","loc_20","loc_21"],[30,17,32,18,17,27,16,1,12,1,27,32,32,13,30,17,29,18,1,11,27],[0.0968672,0.262717,0.530699,0.206103,0.742678,0.00661717,0.781845,0.364942,0.809506,0.297098,0.357502,0.305792,0.445269,0.274888,0.0372731,0.440049,0.789316,0.83912,0.88079,0.23763,0.508397],[0.585599,0.80014,0.90607,0.691145,0.942216,0.5113,0.914278,0.814296,0.933144,0.797127,0.781793,0.8219,0.823699,0.736113,0.559808,0.816611,0.949545,0.948082,0.946734,0.68028,0.852955],[4990.01,362316,99365.2,424394,137621,3309.53,777034,295211,9437.31,121673,27468.2,77185.6,254436,201895,19806.9,491540,192891,287840,403464,73325,191839],[45861.8,1016790,85583.8,1634740,5283.38,496833,216813,236174,1631.14,112032,49365.4,150467,154067,224822,511591,625472,51486.3,55186.1,54606.6,234752,177963],[15.337,29.7177,19.8099,15.8955,3.2863,15.379,6.45877,6.18979,4.0795,16.3237,15.2538,19.1919,6.64464,13.7471,26.2133,8.15475,6.65219,8.50553,2.59966,16.2024,13.6413],[177.42,1018.22,339.462,1399.11,288.059,158.381,893.697,881.821,82.7052,595.026,294.427,440.321,800.052,886.222,365.193,1129.92,299.478,373.609,452.047,592.451,621.256],[53633.7,2080370,175984,1969700,268793,430839,1134080,761005,11781.4,403688,81627.2,191795,547291,696808,531687,1106680,209237,326895,531188,310834,336310],[231.589,1442.35,419.504,1403.46,518.452,656.383,1064.93,872.356,108.542,635.365,285.705,437.944,739.791,834.75,729.169,1051.99,457.424,571.747,728.826,557.525,579.922],[1.205,2.926,13.23,1.676,245.2,0.9586,7.822,3.105,10.04,2.826,2.544,3.5,4.177,2.034,0.9635,3.126,20.74,21.12,15.73,1.631,4.322],[22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22,22],[0.372173,0.0137998,2.18921e-05,0.146431,2.22045e-16,0.547041,0.000214892,0.00769954,6.01914e-06,0.012839,0.0335911,0.0143328,0.00678574,0.0746733,0.546528,0.0098835,3.09598e-07,4.89225e-08,1.03367e-07,0.153441,0.00266536]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Environment<\/th>\n      <th>sel_model<\/th>\n      <th>heritVC<\/th>\n      <th>heritPEV<\/th>\n      <th>genotypicVar<\/th>\n      <th>residualVar<\/th>\n      <th>CV<\/th>\n      <th>LSD<\/th>\n      <th>Variance<\/th>\n      <th>SD<\/th>\n      <th>WaldStatistic<\/th>\n      <th>WaldDF<\/th>\n      <th>Pvalue<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7,8,9,10,11,12,13]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="models-selection" class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> Models Selection</h2>
<p>The table below presents the final models selected according to the
criteria specified earlier. The model specifications presented here are
just informative, but they provide an idea of the relevance of some
model terms in the particular dataset selected.</p>
<ul>
<li>Spatial Model selector
<ul>
<li><code>Model</code>: A numeric value with the number of the selected
Model from the list of evaluated models.</li>
<li><code>RepRow</code>: If TRUE the factor for random effects of row
(or row within replicated if replicate is available) will be added to
the model.</li>
<li><code>RepCol</code>: If TRUE the factor for random effects of column
(or column within replicated if replicate is available) will be added to
the model.</li>
<li><code>splineRow</code>: If TRUE the factor for random effects of
spline for row together with a fixed effect covariate for row will be
added to the model.</li>
<li><code>splineCol</code>: If TRUE the factor for random effects of
spline for column together with a fixed effect covariate for column will
be added to the model.</li>
<li><code>Resid</code>: A string with the specification of type of
residual terms to fit. Options are: ‘indep’ and ‘ar1’ for independent
errors, and autoregressive of order 1 for both rows and columns.</li>
<li><code>Nugget</code>: If TRUE a nugget random effect (only for
autoregressive errors) will be added to the model.</li>
<li><code>List</code>: A numeric value indicating the list category (or
priority) to be used to subset the list of evaluated models.</li>
</ul></li>
<li>Non-spatial model selector</li>
<li>Fitting Criteria</li>
</ul>
<pre class="r"><code>if (length(select.environment) != 0) {
  create_dt(merge(summS[,c(&quot;Environment&quot;, &quot;sel_model&quot;)], 
                  mod.list, 
                  by.x = &quot;sel_model&quot;, 
                  by.y = &quot;Model&quot;, 
                  sort = FALSE))
} else {
  warning(&quot;There is no spatial data!&quot;)
}</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-bbc5385cac48acc15225" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-bbc5385cac48acc15225">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"],[30,30,17,17,17,32,32,32,18,18,27,27,27,16,1,1,1,12,13,29,11],["loc_01","loc_15","loc_02","loc_16","loc_05","loc_03","loc_12","loc_13","loc_18","loc_04","loc_06","loc_11","loc_21","loc_07","loc_10","loc_08","loc_19","loc_09","loc_14","loc_17","loc_20"],[false,false,false,false,false,false,false,false,true,true,false,false,false,false,false,false,false,false,false,false,false],[false,false,false,false,false,true,true,true,false,false,false,false,false,true,false,false,false,false,true,true,false],[true,true,false,false,false,true,true,true,false,false,true,true,true,true,false,false,false,false,true,true,true],[true,true,false,false,false,true,true,true,false,false,false,false,false,true,false,false,false,true,false,false,false],["ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","ar1","indep","indep","indep","indep","indep","indep","ar1","indep"],[false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false],[2,2,2,2,2,1,1,1,1,1,2,2,2,1,2,2,2,2,1,1,2]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sel_model<\/th>\n      <th>Environment<\/th>\n      <th>RepRow<\/th>\n      <th>RepCol<\/th>\n      <th>splineRow<\/th>\n      <th>splineCol<\/th>\n      <th>Resid<\/th>\n      <th>Nugget<\/th>\n      <th>List<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[1,9]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="obtaining-predictions" class="section level2" number="4.6">
<h2><span class="header-section-number">4.6</span> Obtaining
Predictions</h2>
<p>This table is the main output table generated from the first step
analyses, where the adjusted means (BLUEs), are obtained from the first
step analysis, they need to be stored for further use, as these
single-site analyses will not need to be redone in the future.</p>
<ul>
<li><code>gen</code>: Genotype name.</li>
<li><code>predicted.value</code>: Predicted response value (BLUEs).</li>
<li><code>std.error</code>: Standard Error of predicted value.</li>
<li><code>status</code>: A comment to indicate if this predicted value
is estimable.</li>
<li><code>weight</code>: The inverse of standard error.</li>
<li><code>Environment</code>: Environment name.</li>
</ul>
<pre class="r"><code>create_dt(predS)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-22250da24f91f680e4a4" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-22250da24f91f680e4a4">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480","481","482","483"],["geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23","geno_01","geno_02","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_11","geno_12","geno_13","geno_14","geno_15","geno_16","geno_17","geno_18","geno_19","geno_20","geno_21","geno_22","geno_23"],[1224.54,1230.84,1493.42,1443.01,1306.27,1315.07,1349.82,1881.1,1349.19,1212.64,1353.74,1446.1,1272.05,1501.3,1434.62,1269.67,1513.07,1269.51,1597.05,1420.59,1477.18,1188.53,1091.67,2821.3,2335.4,3456.35,3036.56,3823.78,2498.66,3596.86,4038.33,2872.54,4342.62,2879.03,5173.43,4328.75,3299.71,3802.49,2898.82,4205.5,3688.71,2769.55,3486.72,2751.7,3750.04,1769.5,1711.06,758.61,2000.23,1248.1,2039.86,1596.34,1413.66,1676.41,1524.18,1588.75,1015.34,1930.91,1616.18,1646.64,1197.96,1182.66,683.03,1402.41,1793.49,1465.73,1408.46,1448.48,1223.44,8829.5,8989.52,7911.03,9050.35,9032.93,8474.2,7662.65,7639.7,9061.52,7540.63,9365.61,7765.61,8357.76,9485.55,6419.36,7487.39,9150.65,7175.62,5991.94,6702.7,6998.27,8044.11,7338.98,2654.91,2327.02,2373,1810.88,2576.08,2111.64,2047.96,2414.78,1948.05,2351.63,2222.28,1789.89,1771.11,1599.61,2821.88,2524.11,2666.17,2567.4,2200.01,1807.03,2137.51,1422.88,2293.4,4327.96,5089.47,4544.88,4883.28,5154.84,4165.79,3934.87,4452.52,4757.42,5107.73,4013.61,4621.39,4882.98,5009.95,4554.77,4104.86,3786.63,4639.71,4032.98,5085.61,5084.72,4596.42,4653.6,6263.6,5131.33,7346.17,6378.58,8825.42,7816.64,6496.83,7991.11,6267.89,7566.75,6195.14,7888.19,7993.86,7576.08,8269.11,7724.33,8663.11,7204.11,7436.58,6338.36,6299.33,6261.5,7783.3,8093.92,7869.45,7610.22,6982.15,8382.86,7973.46,7716.03,7949.76,6904,8659.93,8540.8,8208.33,7409.13,8575.15,9212.22,7879.35,6950.38,7872.84,8172.75,7761.28,6355.09,8150.12,7430.79,939.347,1011.45,833.828,879.213,1015.58,759.157,876.794,892.441,1046.45,1095.32,1101.13,921.273,936.605,916.157,1021.71,1073.92,1079.74,1116.11,1027.05,1148.12,1132.53,990.695,953.889,1959.66,2085.1,2109.87,1671.55,1508.25,1907.42,1845.74,1808.47,1583.5,2599.68,2698.22,1888.02,2025.58,2255.02,2894.01,1982.97,2579.39,1914.59,1741.66,1842.6,1473.52,1858.81,2938.3,1275.71,1521.27,1219.97,1621.65,1180.78,1915.48,1493.59,1629.71,1396.88,1756.31,1876.66,1271.12,1286.81,1506.33,1342.29,1371.76,1285.37,1222.85,1322.12,1476.21,1307.34,1836.91,1334.42,1970.07,1895.15,2675.89,2458.24,2067.06,1680.67,1674.82,2761.89,1572.68,1924.44,1646.97,2137.74,2172.36,2077.72,2175.77,1816.76,2505.45,2215.81,1898.84,1990.11,1953.89,1799.05,1513.86,5729.2,4427.32,7028.33,5440.52,6360.24,5801,6516.87,5691.34,5862.98,5989.33,5192.13,6601.55,5669.45,5626.97,6519.88,5633.42,5775.52,6746.74,6276.94,5902.45,5833.33,5113.96,6446.07,3708.53,3972.17,3576.32,3066.22,3212.28,3048.82,3636.28,3738.43,3779.33,3555.87,5049.76,3556.05,3661.81,3377.06,4973.15,2991.11,3530.43,3042.84,2565.37,2472.67,2568.42,3295.1,2972.46,2434.54,1890.09,2993.67,2903.19,2403.79,2449.44,3079.55,2748.39,2738.8,3122.18,3060.05,2825.18,2295.73,2768.53,3123.64,2968.97,2536.64,2666.05,3076.77,2572.45,2870.82,3360.24,2158.25,9090.65,8803.1,10693,7193.34,10816.4,8894.17,10220.8,9464.85,10640.6,9410.34,9720.06,9453.23,10049.3,10423.4,10984.6,8977.98,10375.6,9694.26,9908.14,10382,8886,9377.08,9722.07,3527.05,3098.9,3876.99,3980.29,3714.91,2908.2,3563.46,3854.25,2777.35,3480.73,3044.14,3752.79,2769.37,3078.37,3876.29,3745.96,3139,3434.81,3223.07,4331.51,3591.81,3254.13,2419.07,3900.19,1990.96,2560.39,3546.47,2227.62,1814.81,2561.59,2964.2,2334.35,2567.68,2688.67,3015.54,2823.33,2540.47,1886.12,2997.69,3241.04,3370.23,3059.2,2151.15,3516.57,3043.29,2756.21,9878,8429,8742.5,7514,8957.57,10029,8301.5,8947,9723,9900,9472.5,9473,8045.5,9024,9531.5,9427.5,8995,9061,8866.5,8526.5,8479,8122.5,9296,2998.23,3188.24,2826.74,3245.29,2631.29,2787.79,3286.73,3447.23,3513.24,2283.29,3157.79,2322.79,3200.21,3504.21,3862.26,2688.26,2548.76,3196.76,2866.21,3762.21,2494.26,2262.76,3000.21,3307.23,2252.29,3509.61,2010.02,3111.06,3404.81,3208.16,3225.5,2662.43,3258.37,3053.54,2300.35,3342.16,3247.56,3469.21,3090.77,3445.68,4260.96,2889.79,2525.39,3217.27,3024.04,2684.85],[160.697,155.326,155.335,154.02,155.367,153.481,151.653,150.454,155.083,155.518,173.24,169.649,183.471,159.283,162.999,150.726,159.976,149.523,159.068,155.028,158.421,155.296,176.55,619.007,594.179,554.549,556.996,557.629,559.488,559.369,556.015,554.5,550.592,549.938,571.644,625.624,545.616,567.628,550.998,570.033,547.054,567.312,547.308,546.234,547.675,596.979,161.991,157.666,166.654,139.036,164.578,152.846,140.767,155.002,158.404,133.045,203.42,158.333,181.813,157.12,160.047,127.877,137.924,118.909,154.231,135.046,135.712,125.695,205.791,872.743,871.378,860.762,860.39,850.561,850.998,863.491,861.458,858.301,850.74,884.873,861.402,861.788,853.96,851.403,842.02,840.977,853.832,853.985,849.185,839.063,851.894,879.298,81.7259,81.4422,78.0466,81.2608,79.341,77.5631,76.7661,79.489,76.3518,76.2106,76.9885,76.4946,81.3109,84.4202,77.2815,79.6674,78.2454,74.3062,75.6029,80.5307,74.2128,77.4159,76.6807,486.486,480.698,468.171,486.023,490.781,484.973,467.976,469.543,466.129,481.276,477.985,485.932,516.935,480.114,471.395,468.352,471.319,469.789,483.163,475.008,463.474,469.914,491.713,488.851,331.52,338.235,344.78,349.602,344.126,331.52,330.262,330.262,344.271,359.673,343.957,344.127,348.095,330.262,331.52,341.421,330.262,349.602,344.126,338.235,329.252,343.957,431.553,424.962,424.349,424.874,424.587,424.648,424.874,424.357,424.962,424.357,424.587,424.977,431.915,424.977,424.357,424.587,424.357,424.648,431.467,424.587,431.467,431.915,431.553,32.3724,31.9312,32.0189,31.9074,32.0385,31.8777,31.9282,31.9556,32.0273,31.8739,32.7731,31.8997,32.0652,32.1442,32.1152,31.8902,32.79,31.8776,32.1359,31.8853,32.3118,32.066,32.076,307.433,302.058,301.467,301.965,301.7,301.746,301.965,301.475,302.058,301.475,301.7,302.075,307.777,302.075,301.475,301.7,301.475,301.746,307.342,301.7,307.342,307.777,307.433,147.624,144.068,138.118,137.401,138.091,140.627,138.691,137.792,140.183,143.359,151.574,150.335,146.521,140.016,139.083,134.833,140.909,137.425,143.455,141.112,135.238,141.571,153.153,263.127,246.486,232.59,234.193,234.211,228.123,230.929,232.135,233.738,230.876,228.665,237.759,262.5,224.404,233.305,228.648,241.365,225.731,235.173,232.871,226.641,222.81,249.623,393.744,376.505,355.894,375.42,373.158,362.881,347.568,365.457,363.452,363.332,376.181,393.128,438.598,383.323,381.665,366.393,385.702,345.701,372.072,375.711,359.806,356.886,392.675,494.186,496.478,475.687,511.486,498.616,507.343,475.582,495.375,486.916,494.505,490.374,505.971,520.965,514.793,487.404,496.203,487.072,470.685,498.31,498.192,480.448,478.381,484.21,458.661,445.03,434.024,417.893,410.281,413.511,433.941,426.63,442.816,419.018,488.482,446.452,450.662,415.957,429.96,418.486,445.049,398.094,430.834,414.278,416.243,390.441,472.044,539.523,526.772,504.035,499.456,498.896,499.835,498.96,500.242,498.284,502.506,503.917,515.829,541.566,504.751,516.623,504.488,515.111,502.156,515.87,502.967,503.423,502.069,527.901,120.838,121.402,122.571,115.854,127.485,124.665,112.172,125.201,117.703,111.537,135.116,117.574,126.988,118.34,111.475,101.657,102.855,97.9781,121.964,112.364,103.365,96.5417,136.744,153.935,154.681,151.33,151.61,147.67,147.859,153.186,151.471,150.831,148.241,161.232,151.103,150.361,149.702,147.425,145.285,144.637,147.903,148.991,147.363,144.289,148.032,157.467,165.237,165.237,165.237,165.237,236.321,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,165.237,351.88,351.88,351.88,359.455,359.455,359.455,351.88,351.88,351.88,359.455,359.455,359.455,359.455,359.455,342.756,342.756,342.756,342.756,359.455,359.455,342.756,342.756,359.455,310.17,286.345,294.758,260.903,269.777,268.043,292.812,286.92,298.276,279.732,299.184,286.61,270.769,266.037,254.373,244.407,256.65,251.646,265.713,261.27,248.431,261.475,297.284],["Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable"],[4.27529e-05,4.49668e-05,4.38714e-05,4.4964e-05,4.38842e-05,4.53306e-05,4.54387e-05,4.56464e-05,4.47596e-05,4.47078e-05,3.93695e-05,4.08881e-05,3.62272e-05,4.37561e-05,4.22674e-05,4.61334e-05,4.20634e-05,4.64782e-05,4.33046e-05,4.51024e-05,4.2885e-05,4.54298e-05,3.90771e-05,4.67071e-06,5.59078e-06,5.59078e-06,5.59116e-06,5.59116e-06,5.59116e-06,5.59078e-06,5.59078e-06,5.59078e-06,5.59116e-06,5.59028e-06,4.67115e-06,3.16623e-06,5.59116e-06,4.67078e-06,5.59154e-06,4.67166e-06,5.59154e-06,4.9782e-06,5.59116e-06,5.56103e-06,5.59154e-06,3.77926e-06,8.36307e-05,9.52182e-05,9.29423e-05,0.000103351,0.000103306,0.000104289,9.67571e-05,9.73817e-05,9.46618e-05,0.000103418,6.30949e-05,8.92724e-05,8.22191e-05,0.000104039,0.000109636,0.000153655,0.000134734,0.000153699,0.000100719,0.000103944,0.000116072,0.000153565,8.60797e-05,1.48958e-06,1.52519e-06,1.52519e-06,1.61162e-06,1.61162e-06,1.61162e-06,1.52519e-06,1.52519e-06,1.52519e-06,1.61162e-06,1.5027e-06,1.50254e-06,1.50348e-06,1.61162e-06,1.58309e-06,1.70231e-06,1.66195e-06,1.70231e-06,1.57909e-06,1.61162e-06,1.80752e-06,1.70231e-06,1.53564e-06,0.00109564,0.00122587,0.00144434,0.00147016,0.00147214,0.00225801,0.0018303,0.00151567,0.00150445,0.00179312,0.00138088,0.00144501,0.00086853,0.000915407,0.00159638,0.00218425,0.00154399,0.00218436,0.00205248,0.0014727,0.00272309,0.00262114,0.00139866,5.04952e-06,5.36689e-06,5.36689e-06,5.34598e-06,5.34598e-06,5.34598e-06,5.36689e-06,5.36689e-06,5.36689e-06,5.34598e-06,5.22776e-06,4.93864e-06,4.26714e-06,5.34598e-06,5.12808e-06,5.52128e-06,5.19235e-06,5.52128e-06,5.04309e-06,5.34598e-06,5.80253e-06,5.52129e-06,4.58586e-06,4.31967e-06,9.13901e-06,8.88245e-06,8.7871e-06,8.60502e-06,8.80979e-06,9.13902e-06,9.18653e-06,9.18652e-06,8.79814e-06,8.15873e-06,8.81344e-06,8.80974e-06,8.63814e-06,9.18651e-06,9.13902e-06,8.75892e-06,9.18653e-06,8.60502e-06,8.80979e-06,8.88246e-06,9.22453e-06,8.81344e-06,6.49491e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.49491e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.72262e-06,6.49491e-06,6.72262e-06,6.49491e-06,6.49491e-06,6.49491e-06,0.00101223,0.00104394,0.001041,0.0010404,0.00103843,0.00104492,0.00104394,0.00103521,0.00103521,0.0010446,0.00100062,0.00104249,0.00102871,0.00102805,0.0010352,0.00104394,0.000992736,0.0010446,0.0010193,0.00104492,0.00102104,0.00102832,0.00102854,1.34711e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.34711e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.40023e-05,1.34711e-05,1.40023e-05,1.34711e-05,1.34711e-05,1.34711e-05,5.77905e-05,6.20156e-05,6.20156e-05,6.30763e-05,6.30763e-05,6.30763e-05,6.20156e-05,6.20156e-05,6.20156e-05,6.30763e-05,5.60977e-05,5.62921e-05,5.27548e-05,6.30762e-05,6.14782e-05,6.9269e-05,6.46206e-05,6.9269e-05,6.03622e-05,6.30763e-05,7.59609e-05,6.9269e-05,5.57681e-05,2.57299e-05,3.22841e-05,3.22873e-05,3.00524e-05,3.01931e-05,3.4659e-05,3.44956e-05,3.22322e-05,3.22246e-05,3.23679e-05,3.40727e-05,2.8369e-05,1.89791e-05,3.46328e-05,2.85934e-05,3.26958e-05,2.59458e-05,3.26843e-05,3.08606e-05,3.01657e-05,3.08147e-05,3.493e-05,2.28629e-05,9.90021e-06,1.11103e-05,1.10647e-05,1.03912e-05,1.0387e-05,1.17473e-05,1.17818e-05,1.10778e-05,1.10643e-05,1.10391e-05,9.90126e-06,9.55279e-06,8.14443e-06,1.16077e-05,1.03346e-05,1.11452e-05,9.79475e-06,1.11618e-05,1.08507e-05,1.04352e-05,1.06644e-05,1.172e-05,9.13861e-06,5.1446e-06,5.25327e-06,5.25327e-06,5.1513e-06,5.1513e-06,5.1513e-06,5.25327e-06,5.25327e-06,5.22247e-06,5.10955e-06,4.75489e-06,4.79606e-06,5.02205e-06,5.1513e-06,5.22248e-06,5.25327e-06,5.25327e-06,5.22248e-06,5.02205e-06,5.1513e-06,5.11411e-06,5.1446e-06,5.02205e-06,6.37531e-06,6.97652e-06,6.92683e-06,7.74569e-06,7.69448e-06,7.74595e-06,6.97652e-06,6.98572e-06,6.98572e-06,7.7434e-06,6.21744e-06,6.6384e-06,6.13417e-06,7.65512e-06,7.21665e-06,8.52793e-06,7.38424e-06,8.54792e-06,7.24316e-06,7.74475e-06,7.15169e-06,8.56391e-06,6.58971e-06,4.29166e-06,4.66575e-06,4.66575e-06,4.66859e-06,4.66859e-06,4.66859e-06,4.66575e-06,4.66575e-06,4.66575e-06,4.66859e-06,4.66364e-06,4.29145e-06,3.61408e-06,4.66859e-06,4.29289e-06,4.6717e-06,4.29652e-06,4.6717e-06,4.34578e-06,4.66859e-06,4.61068e-06,4.6717e-06,3.93615e-06,0.000111479,0.000120885,0.000120885,0.000129716,0.000129716,0.000129716,0.000120885,0.000120885,0.000120885,0.000129716,0.000100626,0.000115566,0.000115116,0.000129716,0.0001443,0.000186525,0.000171701,0.000186525,0.000127738,0.000129716,0.000150361,0.000186525,0.000118948,5.32281e-05,5.49488e-05,5.49488e-05,6.15431e-05,6.15431e-05,6.15431e-05,5.49488e-05,5.49488e-05,5.49488e-05,6.15431e-05,5.30046e-05,5.4914e-05,5.60227e-05,6.15431e-05,6.08472e-05,6.85017e-05,6.63074e-05,6.85017e-05,5.99937e-05,6.1543e-05,7.46989e-05,6.85017e-05,5.75514e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,1.79058e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,3.66256e-05,8.21317e-06,8.21318e-06,8.21318e-06,8.06e-06,8.06e-06,8.06e-06,8.21318e-06,8.21318e-06,8.21318e-06,8.06e-06,8.06e-06,8.06e-06,8.04304e-06,8.04304e-06,8.5146e-06,8.5146e-06,8.5146e-06,8.5146e-06,8.04304e-06,8.04304e-06,8.5146e-06,8.5146e-06,8.04304e-06,1.5201e-05,1.56442e-05,1.56442e-05,1.80519e-05,1.80519e-05,1.80519e-05,1.56442e-05,1.56442e-05,1.56442e-05,1.80519e-05,1.59222e-05,1.63397e-05,1.64042e-05,1.78411e-05,1.90041e-05,2.11845e-05,2.01899e-05,2.11845e-05,1.77931e-05,1.78411e-05,2.38984e-05,2.11845e-05,1.67149e-05],["loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_01","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_06","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_09","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_15","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_17","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>gen<\/th>\n      <th>predicted.value<\/th>\n      <th>std.error<\/th>\n      <th>status<\/th>\n      <th>weight<\/th>\n      <th>Environment<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="bms-output" class="section level2" number="4.7">
<h2><span class="header-section-number">4.7</span> BMS Output</h2>
<pre class="r"><code>folder.name &lt;- paste0(&quot;BMS_&quot;, format(Sys.Date(), &quot;%Y-%m-%d&quot;), &quot;T&quot;, format(Sys.time(), &quot;%H-%M-%S&quot;))

if (&quot;TRIAL_INSTANCE&quot; %in% colnames(data)) {
  dir.create(paste0(&quot;./outputs/&quot;, folder.name))
  
  col.mean &lt;- paste0(trait.name, &quot;_Means&quot;)
  col.se   &lt;- paste0(trait.name, &quot;_UnitErrors&quot;)
  
  BMSOutput &lt;- data.frame(matrix(ncol = 6, nrow = dim(predS)[1]))
  colnames(BMSOutput) &lt;- c(&quot;TRIAL_INSTANCE&quot;, &quot;LOCATION_NAME&quot;, &quot;ENTRY_NO&quot;, &quot;GID&quot;, col.mean, col.se)
  
  lookup.TRIAL_INSTANCE &lt;- unique(data[,c(&quot;trial&quot;, &quot;TRIAL_INSTANCE&quot;)])
  bms.predS  &lt;- merge(predS, lookup.TRIAL_INSTANCE, by.x = &quot;Environment&quot;, by.y = &quot;trial&quot;, all.x = TRUE)
  
  lookup.germplasm &lt;- germplasm.data[,c(&quot;germplasmName&quot;, &quot;germplasmDbId&quot;, &quot;entryNumber&quot;)]
  lookup.germplasm$germplasmName &lt;- gsub(&quot; &quot;,&quot;&quot;,tolower(lookup.germplasm$germplasmName))
  lookup.germplasm &lt;- lookup.germplasm[!duplicated(lookup.germplasm$germplasmName),]
  bms.predS$gen &lt;- gsub(&quot; &quot;,&quot;&quot;,tolower(bms.predS$gen))
  
  bms.predS  &lt;- merge(bms.predS, lookup.germplasm, by.x = &quot;gen&quot;, by.y = &quot;germplasmName&quot;, all.x = TRUE)
  
  bms.predS  &lt;- bms.predS[with(bms.predS, order(Environment, gen)),]
  
  
  BMSOutput$TRIAL_INSTANCE &lt;- bms.predS$TRIAL_INSTANCE
  BMSOutput$LOCATION_NAME  &lt;- bms.predS$Environment
  BMSOutput$ENTRY_NO       &lt;- bms.predS$entryNumber
  BMSOutput$GID            &lt;- bms.predS$germplasmDbId
  BMSOutput[,c(col.mean)]  &lt;- bms.predS$predicted.value
  BMSOutput[,c(col.se)]    &lt;- bms.predS$std.error
  
  write.csv(BMSOutput, paste0(&quot;./outputs/&quot;, folder.name, &quot;/BMSOutput.csv&quot;), row.names = FALSE)
  
  BMSSummary &lt;- data.frame(matrix(ncol = 28, nrow = dim(summS)[1]))
  colnames(BMSSummary) &lt;- c(&quot;TRIAL_INSTANCE&quot;, &quot;LOCATION_NAME&quot;, &quot;Trait&quot;, &quot;NumValues&quot;, 
                            &quot;NumMissing&quot;, &quot;Mean&quot;, &quot;Variance&quot;, &quot;SD&quot;, &quot;Min&quot;, &quot;Max&quot;, &quot;Range&quot;, 
                            &quot;Median&quot;, &quot;LowerQuartile&quot;, &quot;UpperQuartile&quot;, &quot;MeanRep&quot;, &quot;MinRep&quot;, 
                            &quot;MaxRep&quot;, &quot;MeanSED&quot;, &quot;MinSED&quot;, &quot;MaxSED&quot;, &quot;MeanLSD&quot;, &quot;MinLSD&quot;, 
                            &quot;MaxLSD&quot;, &quot;CV&quot;, &quot;Heritability&quot;, &quot;WaldStatistic&quot;, &quot;WaldDF&quot;, &quot;Pvalue&quot;)
  
  bms.summS  &lt;- merge(summS, lookup.TRIAL_INSTANCE, by.x = &quot;Environment&quot;, by.y = &quot;trial&quot;, all.x = TRUE)
  
  BMSSummary$TRIAL_INSTANCE  &lt;- bms.summS$TRIAL_INSTANCE
  BMSSummary$LOCATION_NAME   &lt;- bms.summS$Environment
  BMSSummary$Trait           &lt;- bms.summS$Trait
  BMSSummary$NumValues       &lt;- bms.summS$nonNA_obs
  BMSSummary$NumMissing      &lt;- bms.summS$NumMissing
  BMSSummary$Mean            &lt;- bms.summS$grandMean
  BMSSummary$Variance        &lt;- bms.summS$Variance
  BMSSummary$SD              &lt;- bms.summS$SD
  BMSSummary$Min             &lt;- bms.summS$Min
  BMSSummary$Max             &lt;- bms.summS$Max
  BMSSummary$Range           &lt;- bms.summS$Range
  BMSSummary$Median          &lt;- bms.summS$Median
  BMSSummary$LowerQuartile   &lt;- bms.summS$LowerQuartile
  BMSSummary$UpperQuartile   &lt;- bms.summS$UpperQuartile
  BMSSummary$MeanRep         &lt;- bms.summS$MeanRep
  BMSSummary$MinRep          &lt;- bms.summS$MinRep
  BMSSummary$MaxRep          &lt;- bms.summS$MaxRep
  BMSSummary$MeanSED         &lt;- merge(bms.summS$Environment, 
                                      aggregate(std.error~Environment, data = predS, mean, na.rm = TRUE), 
                                      by = 1, all.x = TRUE)$std.error
  BMSSummary$MinSED          &lt;- merge(bms.summS$Environment, 
                                      aggregate(std.error~Environment, data = predS, min, na.rm = TRUE), 
                                      by = 1, all.x = TRUE)$std.error
  BMSSummary$MaxSED          &lt;- merge(bms.summS$Environment, 
                                      aggregate(std.error~Environment, data = predS, max, na.rm = TRUE), 
                                      by = 1, all.x = TRUE)$std.error
  BMSSummary$MeanLSD         &lt;- bms.summS$LSD
  BMSSummary$MinLSD          &lt;- merge(bms.summS$Environment, 
                                      aggregate(std.error~Environment, data = predS, min, na.rm = TRUE), 
                                      by = 1, all.x = TRUE)$std.error * 1.96
  BMSSummary$MaxLSD          &lt;- merge(bms.summS$Environment, 
                                      aggregate(std.error~Environment, data = predS, max, na.rm = TRUE), 
                                      by = 1, all.x = TRUE)$std.error * 1.96
  BMSSummary$CV              &lt;- bms.summS$CV
  BMSSummary$Heritability    &lt;- bms.summS$heritVC
  BMSSummary$WaldStatistic   &lt;- bms.summS$WaldStatistic
  BMSSummary$WaldDF          &lt;- bms.summS$WaldDF
  BMSSummary$Pvalue          &lt;- bms.summS$Pvalue
  
  write.csv(BMSSummary, paste0(&quot;./outputs/&quot;, folder.name, &quot;/BMSSummary.csv&quot;), row.names = FALSE)
  
  con &lt;- file(paste0(&quot;./outputs/&quot;, folder.name, &quot;/BMSInformation.txt.txt&quot;))
  writeLines(c(paste0(&quot;InputDataSetId=&quot;),
               paste0(&quot;OutputDataSetId=&quot;, 0),        # this is unused in the system and always defaults to 0
               paste0(&quot;StudyId=&quot;, bms.studyId),
               paste0(&quot;WorkbenchProjectId=&quot;)), con)
  close(con)
  
  setwd(&quot;./outputs&quot;)
  unlink(list.files()[grep(&quot;*\\.zip&quot;, list.files())])
  files2zip &lt;- dir(folder.name, full.names = TRUE)
  zip(zipfile = folder.name, files = files2zip)
  setwd(&quot;..&quot;)
  
  unlink(paste0(&quot;./outputs/&quot;, folder.name), recursive=TRUE)
} else {
  warning(&quot;No BMS TRIAL_INSTANCE data available!&quot;)
}</code></pre>
<blockquote>
<p><em>Get the file to upload BLUEs &amp; summary stats to BMS <a
href="./outputs/BMS_2023-09-06T10-01-37.zip">here</a></em></p>
</blockquote>
</div>
</div>
<div id="met-analysis" class="section level1" number="5">
<h1><span class="header-section-number">5</span> MET Analysis</h1>
<div id="selecting-env.-from-step-1" class="section level2"
number="5.1">
<h2><span class="header-section-number">5.1</span> Selecting Env. from
Step 1</h2>
<p>Before proceeding to step 2, a selection of trials/environments of
interest is performed. In general, this corresponds to all trials
relevant to the current study or analysis. As there might be several
single environment analyses accumulated over time, it is possible to
select these for different objectives. For example, all sites for a
given breeding zone over the last three years. Or trials associated with
some environmental stress (e.g., drought). A key aspect of this
selection is that these trials need to be somehow connected, implying
that there is a need for a group of entries to be common between trials.
This is described in further detail in other sections.</p>
<p>A second criteria is associated with the quality of the information
associated with a given trial/environment. Here, for example, it might
be considered that trials with low heritability (e.g., &lt; 0.10) might
need to be dropped as they do not have sufficient genetic signal to be
considered for the step 2 analysis. Several other criteria can be
considered, such as average yield, residual variance (or CV%), etc. The
aim is to consider a group of trials that should provide reasonable
information.</p>
<p>It is important to note that it is also possible to select only a
relevant set of genotypes; or discard some genotypes that seem poorly
represented or are of no interest. For example, some controls can be
left out if they do not belong to the population of interest. As
indicated earlier, the only critical aspect is that there is sufficient
connectivity between trials in order to have a reasonable estimate of
genetic correlation between trials.</p>
<p>In the case presented below, there are 3 trials that were dropped
given their low heritability (&lt; 0.10), but other criteria might be
considered.</p>
<pre class="r"><code># read stage-1 outputs
summS &lt;- read.csv(paste0(&quot;outputs/&quot;, study.name, &quot;_&quot;, trait.name, &quot;_Stage-1_Summary_Statistics.csv&quot;))
predS &lt;- read.csv(paste0(&quot;outputs/&quot;, study.name, &quot;_&quot;, trait.name, &quot;_Stage-1_Predictions.csv&quot;))

# explore the results from the sites
# eliminate trials with h2VC &lt; threshold (e.g., 0.10)
valid.env &lt;- summS[!is.na(summS$sel_model) &amp; 
                   ((summS$Type != &quot;AUGMENTED&quot; &amp; summS$heritVC &gt; as.numeric(config$met$heritVC.limit)) |
                   (summS$Type == &quot;AUGMENTED&quot; &amp; summS$Pvalue &lt; as.numeric(config$met$Pvalue.limit))), 
                   &quot;Environment&quot;]

# filter to sites of interest/valid
sel.predS &lt;- predS[predS$Environment %in% valid.env,]

excluded.loc  &lt;- trimws(strsplit(config$met$excluded.loc, &quot;,&quot;)[[1]])
excluded.geno &lt;- trimws(strsplit(config$met$excluded.geno, &quot;,&quot;)[[1]])

sel.predS &lt;- sel.predS[!(tolower(sel.predS$Environment) %in% tolower(excluded.loc)),]
sel.predS &lt;- sel.predS[!(tolower(sel.predS$gen) %in% tolower(excluded.geno)),]

case.color &lt;- ifelse(!summS$Environment %in% valid.env | summS$Environment %in% excluded.loc, &quot;coral&quot;, &quot;gray&quot;)

if (summS$Type[1] != &quot;AUGMENTED&quot;) {
  par(mar=c(10,4,1,1))
  barplot(summS$heritVC, ylim = c(0,1), names.arg = summS$Environment, las = 2, 
          ylab = &quot;heritability&quot;, col = case.color)
  abline(h = as.numeric(config$met$heritVC.limit), lty = 2, lwd = 3, col = &quot;red&quot;)
} else {
  par(mar=c(10,4,1,1))
  barplot(summS$Pvalue, ylim = c(0,1), names.arg = summS$Environment, las = 2, 
          ylab = &quot;P-Value&quot;, col = case.color)
  abline(h = as.numeric(config$met$Pvalue.limit), lty = 2, lwd = 3, col = &quot;red&quot;)
}</code></pre>
<p><img src="FAGE_Analysis_Pipeline_files/figure-html/select_environments-1.png" width="672" /></p>
<pre class="r"><code># we may also select/filter the list of geno of interest (e.g. filter out local checks)
message(paste(&quot;&lt;b&gt;Excluded environments in config.ini:&lt;/b&gt;&quot;, config$met$excluded.loc,
              &quot;&lt;br/&gt;&lt;b&gt;Excluded genotypes in config.ini:&lt;/b&gt;&quot;, config$met$excluded.geno))</code></pre>
<div class="alert alert-info">
<p><b>Excluded environments in config.ini:</b> loc_17, loc_09
<br/><b>Excluded genotypes in config.ini:</b> geno_02, geno_11, geno_16,
geno_20</p>
</div>
<p>Excluded sites</p>
<pre class="r"><code>create_dt(summS[!(summS$Environment %in% valid.env), c(&quot;Environment&quot;, &quot;heritVC&quot;, &quot;Pvalue&quot;)])</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-7dc4c2ad12ea91922da6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-7dc4c2ad12ea91922da6">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","6","15"],["loc_01","loc_06","loc_15"],[0.0968672,0.00661717,0.0372731],[0.372173,0.547041,0.546528]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Environment<\/th>\n      <th>heritVC<\/th>\n      <th>Pvalue<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="auditing-data-for-met-analysis" class="section level2"
number="5.2">
<h2><span class="header-section-number">5.2</span> Auditing Data for MET
Analysis</h2>
<p>The selected set of trials/environments to be analyzed, as indicated,
needs to be audited in order to see if this set is reasonable for
downstream analyses. The main aspect is related to connectivity between
trials through genotypes in common. Recall that the <code>corh</code>
and fa structure allow for poor connectivity compared to
<code>corgh</code> and us structures.</p>
<p>If connectivity is poor, it is possible to eliminate some of the
trials that are with limited connections (or no connections). But also,
this might require splitting/separating the group of trials into two or
more sets to be evaluated independently. This can be assessed better
with a heatmap/clustering to separate them. In the example presented
below, all genotypes are present in all sites, therefore this is an
ideal set.</p>
<pre class="r"><code># auditMET object has two components
# $incidence to show/present connectivity between 
# env (in terms of shared geno #), or geno (in terms of shared env #)
# $stats simple summary statistics (n, mean, min, max, sd, and CV)
audit.Env &lt;- auditMET(data=sel.predS, gen=&quot;gen&quot;, trial=&quot;Environment&quot;, resp=&quot;predicted.value&quot;, single=TRUE)
audit.Gen &lt;- auditMET(data=sel.predS, gen=&quot;Environment&quot;, trial=&quot;gen&quot;, resp=&quot;predicted.value&quot;, single=TRUE)

# if there is no variation, then no valid heatmap/clustering can be generated
if (length(table(audit.Env$incidence)) != 1) {
  heatmap(audit.Env$incidence)
  heatmap(audit.Gen$incidence) 
}</code></pre>
<div id="environment-stats" class="section level3" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Environment
Stats</h3>
<p>The table below presents the summary statistics by environment.</p>
<pre class="r"><code>create_dt(audit.Env$stats)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-73db717b81bc9759df21" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-73db717b81bc9759df21">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16"],["loc_02","loc_03","loc_04","loc_05","loc_07","loc_08","loc_10","loc_11","loc_12","loc_13","loc_14","loc_16","loc_18","loc_19","loc_20","loc_21"],[19,19,19,19,19,19,19,19,19,19,19,19,19,19,19,19],[3475.07,1534.19,7996.34,2187.31,7385.69,7821.53,2029.63,1431.87,2065.07,6054.22,3413.94,9752.52,2827.86,8994.03,2962.01,3135.74],[1769.5,683.03,5991.94,1422.88,6261.5,6355.09,1473.52,1180.78,1513.86,5113.96,2565.37,7193.34,1814.81,7514,2262.76,2010.02],[5173.43,2039.86,9485.55,2821.88,8825.42,9212.22,2938.3,1915.48,2761.89,7028.33,4973.15,10984.6,3900.19,10029,3862.26,4260.96],[798.566,320.271,972.93,391.463,839.339,701.252,435.481,220.904,353.16,500.873,528.406,894.518,556.483,687.075,463.023,487.116],[22.9798,20.8756,12.1672,17.897,11.3644,8.96566,21.4561,15.4276,17.1016,8.27311,15.4779,9.17217,19.6786,7.63923,15.632,15.5343]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Environment<\/th>\n      <th>n<\/th>\n      <th>mean<\/th>\n      <th>min<\/th>\n      <th>max<\/th>\n      <th>sd<\/th>\n      <th>CVp<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="genotype-stats" class="section level3" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Genotype Stats</h3>
<p>The table below presents the summary statistics by genotype across
all environment.</p>
<pre class="r"><code>create_dt(audit.Gen$stats)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-c70d45cc60e33430cc67" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-c70d45cc60e33430cc67">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19"],["geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23"],[16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16],[4636.99,4727.48,4142.12,4797.09,4484.68,4480.01,4711.79,4477.95,4705.98,4673.49,4609.58,4735.35,4960.11,4787.57,4664.82,4359.94,4105,4300.66,4406.49],[1275.71,1219.97,1248.1,1180.78,1596.34,1413.66,1629.71,1396.88,1588.75,1271.12,1286.81,1506.33,1197.96,683.03,1222.85,1322.12,1307.34,1422.88,1223.44],[9878,10693,9050.35,10816.4,10029,10220.8,9464.85,10640.6,9900,9473,10049.3,10423.4,10984.6,10375.6,9694.26,9908.14,8886,9377.08,9722.07],[2919.97,2959.53,2539.16,3311.49,3091.54,2870.96,2778.28,3171.02,2974.73,3044.08,2878.6,3187.34,3150,3092.56,2828.76,2905.86,2603.44,2794.04,3018.88],[62.9713,62.6028,61.3011,69.0313,68.9356,64.0838,58.9644,70.814,63.2117,65.1351,62.4482,67.3096,63.5067,64.5957,60.6403,66.649,63.4212,64.9677,68.5097]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>gen<\/th>\n      <th>n<\/th>\n      <th>mean<\/th>\n      <th>min<\/th>\n      <th>max<\/th>\n      <th>sd<\/th>\n      <th>CVp<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<blockquote>
<p>What to do if problems.</p>
</blockquote>
</div>
</div>
<div id="model-fitting-process-with-fax" class="section level2"
number="5.3">
<h2><span class="header-section-number">5.3</span> Model Fitting Process
with FAx</h2>
<p>The key on the step 2 from the two-stage analyses is on fitting a MET
model that has a complex variance-covariance structure for the term
genotype-nested-within-environment factor. This MET model is fitted by
fixing the error variances (single term fixed to 1), but by considering
for each prediction value its corresponding weight.</p>
<p>Several structures are available, but the most relevant one is factor
analytic (FA), for which FA models from order 1 to order 9 are
available. As the order increases, more variance components are
estimated approximating better to the unstructured (<code>us</code>)
matrix structure. However, the order will depend on the number of trials
considered in the set selected.</p>
<p>The table below presents summary and goodness-of-fit statistics for
these models. As these step 2 models only differ on their random
effects, and all these models are nested with each other, the criteria
to use can be either logL, AIC, or BIC. The recommendation is to use the
one with the lowest logL value. But it is also possible to use
sequential likelihood-ratio tests (LRT) to identify the best model. In
any case, once a model is selected it is strongly recommended to assess
the corresponding variance-covariance matrix (as a heatmap of genetic
correlation for example as shown later) to assess its quality (and see
if is biologically reasonable). In this example, an FA of order 6 was
selected.</p>
<pre class="r"><code>faModels &lt;- list()
i &lt;- 1

repeat{
  fa.model &lt;- paste0(&quot;fa&quot;, i)
  
  try({
    faMET &lt;- stageMET(data=sel.predS, gen=&quot;gen&quot;, trial=&quot;Environment&quot;, resp=&quot;predicted.value&quot;, weight=&quot;weight&quot;,
                      type.gen=&quot;random&quot;, type.trial=&quot;fixed&quot;, vc.model=fa.model, rotation=&quot;varimax&quot;, vcovp=&quot;TRUE&quot;)
  
    # check what output components are available
    # ls(faMET)
    faModels[[i]] &lt;- faMET
  
    # good of fit indices and fa cum.var (can be used for criteria to exit the for loop)
    if(i &gt; 1){
      # A test statistic equal to twice the absolute difference in these log-likelihoods
      # is assumed to be distributed as Chi square with one degree of freedom (one term 
      # of difference between the two models).
      # p.value &lt;- 1 - pchisq(2 * (model1$loglik - model2$loglik), 1)
      # if p.value &lt; 0.05 we would therefore conclude that the new model significantly
      # improves the fit of the previous model, given an increase in log-likelihood.
      testing.significance &lt;- signif(1 - pchisq(2 * (as.data.frame(faMET$gfit)$logL - as.numeric(exit.criteria[i-1,&#39;logL&#39;])), 1), digits = 3)
      
      exit.criteria &lt;- rbind(exit.criteria, c(faMET$gfit, tail(faMET$cum.var,1)[2], fa.model, testing.significance))
    }else{
      exit.criteria &lt;- as.data.frame(faMET$gfit)
      exit.criteria$cum.var &lt;- tail(faMET$cum.var,1)[2]
      exit.criteria$fa &lt;- fa.model
      exit.criteria$p.value &lt;- 0
    }
  })
  
  i &lt;- i + 1
  
  if(i &gt; 9) break
}

exit.criteria$logL &lt;- as.numeric(exit.criteria$logL)
exit.criteria$AIC  &lt;- as.numeric(exit.criteria$AIC)
exit.criteria$BIC  &lt;- as.numeric(exit.criteria$BIC)</code></pre>
<div id="how-best-model-is-selected" class="section level3"
number="5.3.1">
<h3><span class="header-section-number">5.3.1</span> How ‘Best’ Model is
Selected</h3>
<p><code>exit.criteria</code></p>
<pre class="r"><code>create_dt(exit.criteria)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-4ce392b97846a134ba22" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4ce392b97846a134ba22">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9"],["33","49","65","81","97","113","129","145","161"],[-1952.6,-1942.11,-1936.6,-1930.74,-1917.99,-1913.45,-1914.16,-2227.12,-2288.43],[3967.2,3974.23,3989.2,3999.49,3999.98,4004.9,4032.31,4658.23,4796.87],[4080.75,4139.06,4201.65,4252.23,4300.34,4330.91,4405.94,5031.86,5199.8],["21.427","46.927","58.177","67.543","75.141","80.948","85.384","99.998","99.998"],["fa1","fa2","fa3","fa4","fa5","fa6","fa7","fa8","fa9"],["0","4.67e-06","0.000898","0.00062","4.41e-07","0.00259","1","1","1"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>n.VC<\/th>\n      <th>logL<\/th>\n      <th>AIC<\/th>\n      <th>BIC<\/th>\n      <th>cum.var<\/th>\n      <th>fa<\/th>\n      <th>p.value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[2,3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code># find the min logL (best model criteria)
if (is.na(as.numeric(config$met$best.fa))) {
  i &lt;- which.max(exit.criteria$logL)
} else {
  i &lt;- as.numeric(config$met$best.fa)
}

sel.model &lt;- faModels[[i]]</code></pre>
<blockquote>
<p>Selected best model is 6</p>
</blockquote>
</div>
<div id="interpreting-final-fitted-model" class="section level3"
number="5.3.2">
<h3><span class="header-section-number">5.3.2</span> Interpreting Final
Fitted Model</h3>
<div id="gxe-correlation-matrix" class="section level4"
number="5.3.2.1">
<h4><span class="header-section-number">5.3.2.1</span> GxE correlation
matrix</h4>
<p>The heatmap below represents an illustration of the genetic
correlations between trials/environments based on the selected FA model.
It is important to check the range of correlation estimates. In this
particular example, there are negative and positive values, and this
needs to be verified with the understanding of the set of trials. This
might be an indication of issues (e.g., poor connectivity or trial
installation difficulties).</p>
<p>In the present example, it is also possible to identify 2 or 3 groups
of trials (by checking dendrograms), and this can assist in separating
the trials into additional sets to be analyzed independently or to
assist on the definition of target population environments (TPEs).</p>
<pre class="r"><code># genetic correlation matrix between environments
# sel.model$corr.M
pheatmap(sel.model$corr.M, color = colorRampPalette(RColorBrewer::brewer.pal(n = 7, name = &quot;RdYlBu&quot;))(100))</code></pre>
<p><img src="FAGE_Analysis_Pipeline_files/figure-html/corr_m-1.png" width="1800" /></p>
</div>
<div id="overall-variance-explained" class="section level4"
number="5.3.2.2">
<h4><span class="header-section-number">5.3.2.2</span> Overall Variance
Explained</h4>
<p>The variance components of the final selected FA model, here FA6, can
be processed to obtain the proportion of variance explained from the
complete GxE analysis. In this example, the 6 factors explain 82.1% of
the total GxE variability.</p>
<pre class="r"><code>create_dt(as.data.frame(sel.model$cum.var))</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-a59f53a10c4a1a126648" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a59f53a10c4a1a126648">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["FA1","FA2","FA3","FA4","FA5","FA6"],[18.437,6.238,13.45,10.62,23.077,9.126],[18.437,24.675,38.125,48.745,71.822,80.948]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>var.exp%<\/th>\n      <th>cum.var.exp%<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[1,2]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="variance-explained-by-env." class="section level4"
number="5.3.2.3">
<h4><span class="header-section-number">5.3.2.3</span> Variance
Explained by Env.</h4>
<p>Exploration of the individual variance components for each of the
trials/environments allows for a calculation of the contribution of each
factor to explain the GxE variability but in this case for each of the
trials. This is presented in the table below, and it is noticed that in
all trials but 3, the cumulative variance explained is larger than 70%
with many trials having values close to 100%. This can lead to further
understanding of the main underly environmental factors that affect the
responses within a given trial.</p>
<pre class="r"><code>create_dt(as.data.frame(sel.model$cum.var.site))</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-eea3ad19b2449723381a" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-eea3ad19b2449723381a">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["loc_02","loc_03","loc_04","loc_05","loc_07","loc_08","loc_10","loc_11","loc_12","loc_13","loc_14","loc_16","loc_18","loc_19","loc_20","loc_21"],[58.777,0.067,1.323,2.186,15.128,0.98,0.027,4.86,94.811,0,11.876,0.003,5.114,11.461,0.075,1.763],[0.124,86.762,0.041,3.025,0.916,2.345,60.421,0.127,0.154,4.941,14.058,0.262,1.706,0.132,3.873,2.328],[2.457,3.137,11.179,4.741,2.418,0.152,0.421,80.205,0.047,58.355,1.052,43.519,0.771,0.022,0.834,5.031],[4.341,0.466,19.954,87.355,8.109,2.522,4.812,9.479,3.29,21.168,3.085,0.52,0.196,25.415,0.826,20.723],[0.002,0.22,3.079,2.367,25.126,2.597,7.28,0.123,0.042,12.648,30.199,40.977,73.599,1.718,29.452,3.073],[22.783,9.347,0.311,0.326,15.447,91.404,27.039,5.206,1.655,2.887,39.729,14.72,18.614,26.641,0.405,2.493],[88.484,100,35.887,100,67.143,100,100,100,100,100,100,100,100,65.389,35.466,35.412]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FA1<\/th>\n      <th>FA2<\/th>\n      <th>FA3<\/th>\n      <th>FA4<\/th>\n      <th>FA5<\/th>\n      <th>FA6<\/th>\n      <th>cum.var.exp%<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="predictions" class="section level4" number="5.3.2.4">
<h4><span class="header-section-number">5.3.2.4</span> Predictions</h4>
<p>The table below presents the final predictions for all entries in
each of the trials with their standard error. In addition, there is an
additional column that indicates if the prediction corresponds to an
extrapolation, this is the case for some entries that have a prediction
for trials/environments on which they were not evaluated.</p>
<p>Note that these are the <strong>final</strong> predictions for each
entry in each trial based on the estimated variance-covariance structure
selected. In addition, their standard errors will reflect the level of
information available on each of these predictions. Nevertheless, it is
important to consider these standard errors as approximated (and likely
to be underestimated); however, their predicted values are unbiased.</p>
<p>It is recommended to use this table for calculating predictions
across trials of interest (a subset of the ones here if these correspond
to different target population environments) and to perform any
downstream stability calculations to select the best entries (with broad
or specific adaptability), and calculations of genetic gains.</p>
<pre class="r"><code>create_dt(sel.model$predictions)</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-2da6ea57ac0b4d2a8949" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-2da6ea57ac0b4d2a8949">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304"],["loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_02","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_03","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_04","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_05","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_07","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_08","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_10","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_11","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_12","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_13","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_14","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_16","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_18","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_19","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_20","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21","loc_21"],["geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23","geno_01","geno_03","geno_04","geno_05","geno_06","geno_07","geno_08","geno_09","geno_10","geno_12","geno_13","geno_14","geno_15","geno_17","geno_18","geno_19","geno_21","geno_22","geno_23"],[3107.04,3806.62,3224.12,3653.46,2639.29,3263.23,4045.71,2720.72,3856.58,4484.97,4110.31,3834.07,3899.76,3889.61,3543.04,3124.52,2689.96,3664.57,2569.84,1631.93,1926.68,1268.27,2010.84,1603.43,1452.59,1638.05,1531.08,1567.52,1940.17,1576.15,1637.59,1235.55,745.396,1435.77,1786.54,1416.82,1462.54,1205.57,7614.11,7963.9,8839.58,8060.53,8285.03,7910.77,8158.87,8267.18,7703.81,7927.51,8392.49,8557.43,7363.54,8289.89,7497.26,7315.63,7777.76,8342.03,7535.65,2650.2,2374.32,1810.2,2570.13,2112.08,2046.44,2412.85,1947.92,2349.64,1794.62,1775.6,1609.96,2820.25,2665.28,2568.38,2200.08,2137.11,1423.37,2295.42,6617.8,7584.43,6449.99,8555.36,7642.69,6774.32,7961.38,6534.3,7635.46,7695.38,7794.18,7500.99,8473.6,8400.93,7286.48,7277.89,6351.21,6410.38,7507.31,7893.75,7909.15,6808.01,8264.16,8165.3,7600.65,8049.81,7361.53,8777.21,8363.63,7645.17,8158.11,8732.14,7286.67,7639.76,7817.24,6738.83,7833.07,7560.77,1916.05,1751.37,1617.54,1811.07,1989.56,2073.61,1910.8,1875.94,2369.47,1880.66,1926.86,2073.67,2876.56,2633.85,2106.22,1797.13,1619.22,1985.89,2331.59,1313.32,1176.81,1650.7,1262.74,1838.06,1403.5,1498.79,1445.99,1639.46,1373.89,1386.19,1569.79,1401.44,1298.21,1199.88,1324.13,1330.93,1737.77,1396.51,2084.1,2393.58,2334.08,2245.24,1669.85,1823.1,2600.04,1633.4,1990.4,2170.6,2257.74,1956.86,2126.75,2479.67,2211.11,1910.96,1977.12,1854.44,1590.42,6155.02,6623.66,5210.85,6700.01,5745.5,6142.1,5929.1,6139.29,5937.52,6128.85,5989.01,5846.64,6512.12,5997.1,6338.56,6272.78,5921.99,5337.41,6159.03,3055.03,3523.72,2932.89,3586.67,3415.27,3380.19,3582.25,3188.61,3724.2,3433.69,3481.81,3550.39,4327.55,3839,3379.78,3136.54,2782.85,3223.75,3348.9,8877.12,10616.4,8029.7,10649.2,9391.81,10163.7,9299.11,10125.9,9600.87,9975.44,10060.8,10018.6,11172.1,9907.97,9843.51,9736.95,8946.51,8888.84,10036.8,3769.76,2598.36,3579.02,2266.49,1892.98,2581.79,2939.54,2350.88,2581.6,3005.2,2821.47,2496.9,1935.33,3223.6,3345.15,3033.31,3533.4,3020.19,2743.05,9823.67,8765.43,7581.89,9083.1,9972.64,8393.24,8939.47,9615.1,9876.5,9376.12,8081.53,8979.9,9578.12,8942.04,9061.44,8911.17,8486.18,8180.14,9303.05,2753.36,2988.22,2968.99,2937.62,2973.85,3139.08,3107.72,3247.34,2669.83,2599.64,3064.74,3119.72,3497.88,2984.56,3016.58,2838.98,2743.97,2580.9,3034.17,3304.8,3436.14,2245.57,3200.49,3279.06,3165.55,3228.96,2774.19,3259.47,2537.02,3216.35,3133.35,3535.27,3455.05,4005.49,2962.59,3152.77,2926.38,2861.42],[300.088,283.613,284.47,284.48,282.594,282.686,283.567,283.593,283.518,296.961,326.441,282.811,296.333,297.641,283.077,291.967,284.582,282.601,313.127,95.9201,91.8047,88.0638,88.1169,87.6606,90.3156,90.1332,91.1476,88.0004,93.4096,96.7545,87.7616,86.0018,79.0018,74.6379,89.0855,84.0696,74.7189,94.9363,421.997,419.225,416.355,416.511,415.603,418.849,419.212,419.222,415.967,420.884,423.186,415.715,417.661,415.205,412.506,417.517,409.216,412.243,420.834,29.9576,26.1365,25.9096,25.8961,20.9526,23.2493,25.5219,25.6157,23.4871,26.1355,33.5849,32.6999,24.8785,25.2926,21.2994,21.9693,19.0939,19.4581,26.5635,343.452,276.319,277.359,279.325,276.874,273.543,273.169,273.172,277.117,277.508,278.768,278.756,273.48,278.057,272.97,279.462,276.18,272.503,278.11,246.199,241.56,240.93,241.912,240.825,241.288,241.275,241.451,240.919,243.633,248.695,240.862,241.813,240.591,238.439,244.041,241.359,240.366,247.105,141.234,137.98,136.899,137.093,136.632,137.44,137.449,137.778,136.819,139.328,142.66,136.677,136.851,134.908,132.896,138.539,136.536,133.745,141.431,79.861,77.7183,77.5362,77.625,77.0229,77.4373,77.6995,77.7121,77.2883,80.2634,83.0829,77.1103,78.4513,77.7605,75.6413,78.5209,74.4784,75.4937,81.1437,137.543,127.956,130.308,130.742,125.421,125.633,127.969,127.986,127.788,133.372,148.813,125.56,132.846,136.021,127.257,130.055,129.298,125.005,141.657,140.992,136.813,137.047,137.049,135.126,135.666,136.613,136.735,136.122,141.325,147.345,135.59,138.253,137.647,133.4,137.877,133.657,132.677,143.446,154.142,150.405,150.286,150.298,149.832,150.095,150.278,150.378,150.115,152.739,156.884,149.875,151.394,151.193,148.953,151.699,150.009,149.211,154.867,284.276,275.708,275.308,275.347,273.619,274.801,275.669,275.693,274.498,284.491,297.279,273.818,280.343,279.187,271.243,279.743,269.7,270.714,289.655,126.347,124.148,118.284,118.303,118.255,124.122,124.136,124.139,118.271,124.484,123.848,118.263,118.995,114.624,112.787,119.742,108.566,112.78,122.199,155.661,155.523,155.54,210.27,155.471,155.486,155.519,155.521,155.504,155.591,155.809,155.497,155.56,155.584,155.462,155.542,155.501,155.443,155.702,216.259,216.074,216.68,216.713,216.651,216.042,216.05,216.064,216.667,216.919,217.151,216.745,214.434,214.271,214.127,216.841,214.177,214.135,217.042,209.011,206.978,197.069,197.069,197.055,206.967,206.975,206.976,197.062,203.983,203.771,197.9,193.533,189.375,186.065,198.068,177.897,186.063,202.428],["Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable","Estimable"],[false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>trial<\/th>\n      <th>gen<\/th>\n      <th>predicted.value<\/th>\n      <th>std.error<\/th>\n      <th>status<\/th>\n      <th>extrap<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="rotate-loadings" class="section level4" number="5.3.2.5">
<h4><span class="header-section-number">5.3.2.5</span> Rotate
Loadings</h4>
<p>The estimated matrix of loadings across all trials is often rotated
to facilitate its interpretation. In this case, a <code>varimax</code>
is considered by default. It is possible to associate (correlate) these
rotated loadings to environmental variables by simple linear
relationships for further understanding of the environmental variables
driving GxE.</p>
<pre class="r"><code># n rows for sites, and m columns for fa components
# each fa loading vector may correlated with available environmental variables
# trying to find the highest correlated one that can interpret this component
# biplot technique can be used also to interpret this result
#
# sel.model$fa.loadings
# sel.model$fa.rot.loadings
create_dt(as.data.frame(sel.model$fa.rot.loadings))</code></pre>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-6fcd31ea81c81389f6e7" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-6fcd31ea81c81389f6e7">{"x":{"filter":"none","vertical":false,"extensions":["Buttons"],"data":[["loc_02","loc_03","loc_04","loc_05","loc_07","loc_08","loc_10","loc_11","loc_12","loc_13","loc_14","loc_16","loc_18","loc_19","loc_20","loc_21"],[-482.679,-7.97651,-65.2813,-57.6648,-294.777,-58.4926,-5.6764,43.1612,-304.367,-0.838519,-124.58,-4.42074,-123.343,226.293,-8.45016,-56.9339],[-22.1524,287.244,-11.5491,-67.8436,-72.5248,90.4739,-270.457,-6.9888,12.2809,87.1982,-135.543,-40.2023,-71.2438,24.2771,-60.5634,-65.431],[-98.696,-54.6183,189.756,-84.9317,-117.847,22.9986,-22.5877,175.347,-6.79626,-299.67,-37.0854,-517.839,-47.8773,-9.97604,-28.107,-96.1806],[131.177,21.0602,253.516,-364.55,-215.818,-93.8351,-76.3237,60.2793,-56.6984,-180.487,-63.4918,-56.5847,-24.1424,-336.981,-27.9696,-195.202],[2.50192,14.4679,-99.5766,-60.0044,-379.899,-95.2097,-93.8805,-6.87626,6.39424,-139.514,-198.657,-502.49,467.91,-87.6146,-167.001,-75.1743],[300.515,94.2824,-31.6379,22.2654,297.875,564.865,180.925,44.6748,-40.215,66.651,227.858,301.165,-235.316,345.016,-19.579,67.6978]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FA1<\/th>\n      <th>FA2<\/th>\n      <th>FA3<\/th>\n      <th>FA4<\/th>\n      <th>FA5<\/th>\n      <th>FA6<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"Blfrtip","scrollX":true,"buttons":["copy","csv","excel","pdf","print"],"lengthMenu":[[10,25,50,-1],["10","25","50","All"]],"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="latent-regression-plots" class="section level4"
number="5.3.2.6">
<h4><span class="header-section-number">5.3.2.6</span> Latent Regression
Plots</h4>
<p>The use of FA models with an exploration of the BLUP effects together
with the factor loadings allows for a better understanding of the
dynamics of GxE for a given genotype. The plots presented here can be
assessed for this purpose. For example, finding if the relationship is
flat (no slope) will indicate that the given genotype is not affected by
the underlying factor. In contrast, strong negative or positive slopes
will indicate some GxE response with large or small effects on a set of
trials. A genotype with a projection of zero on the y-axis has an
average response at the given trial.</p>
<pre class="r"><code>latent.reg.plots &lt;- paste(&quot;fa&quot;, expand.grid(1:i, levels(sel.model$predictions$gen))$Var1, 
                          &quot;_&quot;,  expand.grid(1:i, levels(sel.model$predictions$gen))$Var2, 
                          sep = &quot;&quot;)

select.latent.reg &lt;- paste0(&quot;plots/latent_reg_&quot;, latent.reg.plots, &quot;_&quot;, trait.name, &quot;.png&quot;)
names(select.latent.reg) &lt;- latent.reg.plots

bsselect(select.latent.reg, 
         type = &quot;iframe&quot;, 
         live_search = TRUE, 
         show_tick = TRUE, 
         elementId = &#39;latent_reg&#39;)$prepend[[1]]$html</code></pre>
<html>
<select id="pmkXEZJbkc" class="selectpicker" data-dropdown-align-right="false" data-dropup-auto="true" data-header="false" data-live-search="true" data-live-search-style="contains" data-show-tick="true" data-width="false" data-size="auto"><option value="plots/latent_reg_fa1_geno_01_GY_Calc_kgha.png">fa1_geno_01</option>
<option value="plots/latent_reg_fa2_geno_01_GY_Calc_kgha.png">fa2_geno_01</option>
<option value="plots/latent_reg_fa3_geno_01_GY_Calc_kgha.png">fa3_geno_01</option>
<option value="plots/latent_reg_fa4_geno_01_GY_Calc_kgha.png">fa4_geno_01</option>
<option value="plots/latent_reg_fa5_geno_01_GY_Calc_kgha.png">fa5_geno_01</option>
<option value="plots/latent_reg_fa6_geno_01_GY_Calc_kgha.png">fa6_geno_01</option>
<option value="plots/latent_reg_fa1_geno_03_GY_Calc_kgha.png">fa1_geno_03</option>
<option value="plots/latent_reg_fa2_geno_03_GY_Calc_kgha.png">fa2_geno_03</option>
<option value="plots/latent_reg_fa3_geno_03_GY_Calc_kgha.png">fa3_geno_03</option>
<option value="plots/latent_reg_fa4_geno_03_GY_Calc_kgha.png">fa4_geno_03</option>
<option value="plots/latent_reg_fa5_geno_03_GY_Calc_kgha.png">fa5_geno_03</option>
<option value="plots/latent_reg_fa6_geno_03_GY_Calc_kgha.png">fa6_geno_03</option>
<option value="plots/latent_reg_fa1_geno_04_GY_Calc_kgha.png">fa1_geno_04</option>
<option value="plots/latent_reg_fa2_geno_04_GY_Calc_kgha.png">fa2_geno_04</option>
<option value="plots/latent_reg_fa3_geno_04_GY_Calc_kgha.png">fa3_geno_04</option>
<option value="plots/latent_reg_fa4_geno_04_GY_Calc_kgha.png">fa4_geno_04</option>
<option value="plots/latent_reg_fa5_geno_04_GY_Calc_kgha.png">fa5_geno_04</option>
<option value="plots/latent_reg_fa6_geno_04_GY_Calc_kgha.png">fa6_geno_04</option>
<option value="plots/latent_reg_fa1_geno_05_GY_Calc_kgha.png">fa1_geno_05</option>
<option value="plots/latent_reg_fa2_geno_05_GY_Calc_kgha.png">fa2_geno_05</option>
<option value="plots/latent_reg_fa3_geno_05_GY_Calc_kgha.png">fa3_geno_05</option>
<option value="plots/latent_reg_fa4_geno_05_GY_Calc_kgha.png">fa4_geno_05</option>
<option value="plots/latent_reg_fa5_geno_05_GY_Calc_kgha.png">fa5_geno_05</option>
<option value="plots/latent_reg_fa6_geno_05_GY_Calc_kgha.png">fa6_geno_05</option>
<option value="plots/latent_reg_fa1_geno_06_GY_Calc_kgha.png">fa1_geno_06</option>
<option value="plots/latent_reg_fa2_geno_06_GY_Calc_kgha.png">fa2_geno_06</option>
<option value="plots/latent_reg_fa3_geno_06_GY_Calc_kgha.png">fa3_geno_06</option>
<option value="plots/latent_reg_fa4_geno_06_GY_Calc_kgha.png">fa4_geno_06</option>
<option value="plots/latent_reg_fa5_geno_06_GY_Calc_kgha.png">fa5_geno_06</option>
<option value="plots/latent_reg_fa6_geno_06_GY_Calc_kgha.png">fa6_geno_06</option>
<option value="plots/latent_reg_fa1_geno_07_GY_Calc_kgha.png">fa1_geno_07</option>
<option value="plots/latent_reg_fa2_geno_07_GY_Calc_kgha.png">fa2_geno_07</option>
<option value="plots/latent_reg_fa3_geno_07_GY_Calc_kgha.png">fa3_geno_07</option>
<option value="plots/latent_reg_fa4_geno_07_GY_Calc_kgha.png">fa4_geno_07</option>
<option value="plots/latent_reg_fa5_geno_07_GY_Calc_kgha.png">fa5_geno_07</option>
<option value="plots/latent_reg_fa6_geno_07_GY_Calc_kgha.png">fa6_geno_07</option>
<option value="plots/latent_reg_fa1_geno_08_GY_Calc_kgha.png">fa1_geno_08</option>
<option value="plots/latent_reg_fa2_geno_08_GY_Calc_kgha.png">fa2_geno_08</option>
<option value="plots/latent_reg_fa3_geno_08_GY_Calc_kgha.png">fa3_geno_08</option>
<option value="plots/latent_reg_fa4_geno_08_GY_Calc_kgha.png">fa4_geno_08</option>
<option value="plots/latent_reg_fa5_geno_08_GY_Calc_kgha.png">fa5_geno_08</option>
<option value="plots/latent_reg_fa6_geno_08_GY_Calc_kgha.png">fa6_geno_08</option>
<option value="plots/latent_reg_fa1_geno_09_GY_Calc_kgha.png">fa1_geno_09</option>
<option value="plots/latent_reg_fa2_geno_09_GY_Calc_kgha.png">fa2_geno_09</option>
<option value="plots/latent_reg_fa3_geno_09_GY_Calc_kgha.png">fa3_geno_09</option>
<option value="plots/latent_reg_fa4_geno_09_GY_Calc_kgha.png">fa4_geno_09</option>
<option value="plots/latent_reg_fa5_geno_09_GY_Calc_kgha.png">fa5_geno_09</option>
<option value="plots/latent_reg_fa6_geno_09_GY_Calc_kgha.png">fa6_geno_09</option>
<option value="plots/latent_reg_fa1_geno_10_GY_Calc_kgha.png">fa1_geno_10</option>
<option value="plots/latent_reg_fa2_geno_10_GY_Calc_kgha.png">fa2_geno_10</option>
<option value="plots/latent_reg_fa3_geno_10_GY_Calc_kgha.png">fa3_geno_10</option>
<option value="plots/latent_reg_fa4_geno_10_GY_Calc_kgha.png">fa4_geno_10</option>
<option value="plots/latent_reg_fa5_geno_10_GY_Calc_kgha.png">fa5_geno_10</option>
<option value="plots/latent_reg_fa6_geno_10_GY_Calc_kgha.png">fa6_geno_10</option>
<option value="plots/latent_reg_fa1_geno_12_GY_Calc_kgha.png">fa1_geno_12</option>
<option value="plots/latent_reg_fa2_geno_12_GY_Calc_kgha.png">fa2_geno_12</option>
<option value="plots/latent_reg_fa3_geno_12_GY_Calc_kgha.png">fa3_geno_12</option>
<option value="plots/latent_reg_fa4_geno_12_GY_Calc_kgha.png">fa4_geno_12</option>
<option value="plots/latent_reg_fa5_geno_12_GY_Calc_kgha.png">fa5_geno_12</option>
<option value="plots/latent_reg_fa6_geno_12_GY_Calc_kgha.png">fa6_geno_12</option>
<option value="plots/latent_reg_fa1_geno_13_GY_Calc_kgha.png">fa1_geno_13</option>
<option value="plots/latent_reg_fa2_geno_13_GY_Calc_kgha.png">fa2_geno_13</option>
<option value="plots/latent_reg_fa3_geno_13_GY_Calc_kgha.png">fa3_geno_13</option>
<option value="plots/latent_reg_fa4_geno_13_GY_Calc_kgha.png">fa4_geno_13</option>
<option value="plots/latent_reg_fa5_geno_13_GY_Calc_kgha.png">fa5_geno_13</option>
<option value="plots/latent_reg_fa6_geno_13_GY_Calc_kgha.png">fa6_geno_13</option>
<option value="plots/latent_reg_fa1_geno_14_GY_Calc_kgha.png">fa1_geno_14</option>
<option value="plots/latent_reg_fa2_geno_14_GY_Calc_kgha.png">fa2_geno_14</option>
<option value="plots/latent_reg_fa3_geno_14_GY_Calc_kgha.png">fa3_geno_14</option>
<option value="plots/latent_reg_fa4_geno_14_GY_Calc_kgha.png">fa4_geno_14</option>
<option value="plots/latent_reg_fa5_geno_14_GY_Calc_kgha.png">fa5_geno_14</option>
<option value="plots/latent_reg_fa6_geno_14_GY_Calc_kgha.png">fa6_geno_14</option>
<option value="plots/latent_reg_fa1_geno_15_GY_Calc_kgha.png">fa1_geno_15</option>
<option value="plots/latent_reg_fa2_geno_15_GY_Calc_kgha.png">fa2_geno_15</option>
<option value="plots/latent_reg_fa3_geno_15_GY_Calc_kgha.png">fa3_geno_15</option>
<option value="plots/latent_reg_fa4_geno_15_GY_Calc_kgha.png">fa4_geno_15</option>
<option value="plots/latent_reg_fa5_geno_15_GY_Calc_kgha.png">fa5_geno_15</option>
<option value="plots/latent_reg_fa6_geno_15_GY_Calc_kgha.png">fa6_geno_15</option>
<option value="plots/latent_reg_fa1_geno_17_GY_Calc_kgha.png">fa1_geno_17</option>
<option value="plots/latent_reg_fa2_geno_17_GY_Calc_kgha.png">fa2_geno_17</option>
<option value="plots/latent_reg_fa3_geno_17_GY_Calc_kgha.png">fa3_geno_17</option>
<option value="plots/latent_reg_fa4_geno_17_GY_Calc_kgha.png">fa4_geno_17</option>
<option value="plots/latent_reg_fa5_geno_17_GY_Calc_kgha.png">fa5_geno_17</option>
<option value="plots/latent_reg_fa6_geno_17_GY_Calc_kgha.png">fa6_geno_17</option>
<option value="plots/latent_reg_fa1_geno_18_GY_Calc_kgha.png">fa1_geno_18</option>
<option value="plots/latent_reg_fa2_geno_18_GY_Calc_kgha.png">fa2_geno_18</option>
<option value="plots/latent_reg_fa3_geno_18_GY_Calc_kgha.png">fa3_geno_18</option>
<option value="plots/latent_reg_fa4_geno_18_GY_Calc_kgha.png">fa4_geno_18</option>
<option value="plots/latent_reg_fa5_geno_18_GY_Calc_kgha.png">fa5_geno_18</option>
<option value="plots/latent_reg_fa6_geno_18_GY_Calc_kgha.png">fa6_geno_18</option>
<option value="plots/latent_reg_fa1_geno_19_GY_Calc_kgha.png">fa1_geno_19</option>
<option value="plots/latent_reg_fa2_geno_19_GY_Calc_kgha.png">fa2_geno_19</option>
<option value="plots/latent_reg_fa3_geno_19_GY_Calc_kgha.png">fa3_geno_19</option>
<option value="plots/latent_reg_fa4_geno_19_GY_Calc_kgha.png">fa4_geno_19</option>
<option value="plots/latent_reg_fa5_geno_19_GY_Calc_kgha.png">fa5_geno_19</option>
<option value="plots/latent_reg_fa6_geno_19_GY_Calc_kgha.png">fa6_geno_19</option>
<option value="plots/latent_reg_fa1_geno_21_GY_Calc_kgha.png">fa1_geno_21</option>
<option value="plots/latent_reg_fa2_geno_21_GY_Calc_kgha.png">fa2_geno_21</option>
<option value="plots/latent_reg_fa3_geno_21_GY_Calc_kgha.png">fa3_geno_21</option>
<option value="plots/latent_reg_fa4_geno_21_GY_Calc_kgha.png">fa4_geno_21</option>
<option value="plots/latent_reg_fa5_geno_21_GY_Calc_kgha.png">fa5_geno_21</option>
<option value="plots/latent_reg_fa6_geno_21_GY_Calc_kgha.png">fa6_geno_21</option>
<option value="plots/latent_reg_fa1_geno_22_GY_Calc_kgha.png">fa1_geno_22</option>
<option value="plots/latent_reg_fa2_geno_22_GY_Calc_kgha.png">fa2_geno_22</option>
<option value="plots/latent_reg_fa3_geno_22_GY_Calc_kgha.png">fa3_geno_22</option>
<option value="plots/latent_reg_fa4_geno_22_GY_Calc_kgha.png">fa4_geno_22</option>
<option value="plots/latent_reg_fa5_geno_22_GY_Calc_kgha.png">fa5_geno_22</option>
<option value="plots/latent_reg_fa6_geno_22_GY_Calc_kgha.png">fa6_geno_22</option>
<option value="plots/latent_reg_fa1_geno_23_GY_Calc_kgha.png">fa1_geno_23</option>
<option value="plots/latent_reg_fa2_geno_23_GY_Calc_kgha.png">fa2_geno_23</option>
<option value="plots/latent_reg_fa3_geno_23_GY_Calc_kgha.png">fa3_geno_23</option>
<option value="plots/latent_reg_fa4_geno_23_GY_Calc_kgha.png">fa4_geno_23</option>
<option value="plots/latent_reg_fa5_geno_23_GY_Calc_kgha.png">fa5_geno_23</option>
<option value="plots/latent_reg_fa6_geno_23_GY_Calc_kgha.png">fa6_geno_23</option></select>
<iframe src="plots/latent_reg_fa1_geno_01_GY_Calc_kgha.png" frameborder="0" height="500" width="100%" id="bIxvChRtNM"></iframe>
<script>$(document).ready(function(){
                   $("#pmkXEZJbkc").change(function(){
                   $("#bIxvChRtNM").attr("src",$(this).val());

                   });

    });</script>
</html>
<blockquote>
<p><em>You can find the latent regression plots <a
href="./outputs/Study_GY_Calc_kgha_Stage-2_Latent_Reg_Plots_FA.pdf">here</a></em></p>
</blockquote>
</div>
<div id="model-call" class="section level4" number="5.3.2.7">
<h4><span class="header-section-number">5.3.2.7</span> Model Call</h4>
<p>In the present workflow, all analyses are done using ASReml-R. Given
the complexity of some of the models, and future interest in exploring
further the model fitted, the following objects obtained from the MET
fitted in the second step provide the definition of the LMM as required
for ASReml-R.</p>
<ul>
<li><code>call</code>: String with the ASReml-R call used to fit the
requested model.</li>
<li><code>mod</code>: ASReml-R object with all information from the
fitted model.</li>
</ul>
<pre class="r"><code># the whole fitted model still available to extract any further info (e.g., the full list of varcomp)
# summary(sel.model$mod)$varcomp</code></pre>
<ul>
<li><code>vcov.M</code>: GxE variance-covariance matrix.</li>
<li><code>vcov.P</code>: Variance-covariance matrix of predictions.</li>
</ul>
<pre class="r"><code># calculate the genetic variance per environment (i.e., measure genetic variability in each env)
# this can be useful in site selection for example
# sqrt(diag(sel.model$vcov.M))</code></pre>
</div>
</div>
</div>
</div>
<div id="closing-remarks" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Closing Remarks</h1>
<p>The current workflow for MET analyses as illustrated here represents
a very practical and operational procedure to follow based on the
state-of-the-art multi-environmental trials analysis and using the
modern statistical software considering a complex variance-covariance
structure, such as FA. This procedure is backed by several manuscripts
and by the personal experience of the authors.</p>
<p>As with any workflow, this is dynamic and future advancements and use
of it will allow for improvements. However, some of the elements that
are currently identified as relevant in future editions are:</p>
<ul>
<li><p>Incorporation of a pedigree-based or genomic-based relationship
matrix to estimate breeding values, additive x environment interaction,
and narrow-sense and genomic heritabilities.</p></li>
<li><p>Extensions on the graphical output and interpretation of the
Biplots.</p></li>
<li><p>Addition of functions for stability indexes that will assist in
the selection of future entries.</p></li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "FAGE_Analysis_Pipeline_files/mathjax-local/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
